---
title: "Jurkat Stimulated Cell Line Study - Manuscript Figures"
author: "Zain Arifin"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    highlight: zenburn
    theme: simplex
---

```{r setup, include=FALSE}
library(tidyverse)
library(Seurat)
library(data.table)
library(reshape2)
library(ggrepel)
library(DESeq2)
library(ComplexHeatmap)
library(RColorBrewer)
library(msigdbr)
library(fgsea)
library(PPInfer)
library(Mfuzz)
library(cowplot)
library(shiny)
library(UpSetR)
library(VennDiagram)
library(openxlsx)
library(factoextra)
library(FactoMineR)
library(Seurat)
library(gprofiler2)
library(EnhancedVolcano)
library(rstatix)

options(scipen = 999)

normalise_filter_counts <- function(counts){
    return(log2(counts + 1))
}

`%ni%` <- Negate(`%in%`)

left_join_multiple <- function(...){
  df1 = list(...)[[1]]
  df2 = list(...)[[2]]
  col1 = colnames(df1)[1]
  col2 = colnames(df2)[1]
  xxx = dplyr::left_join(..., by = setNames(col2,col1))
  return(xxx)
}

msigdbr_df_H <- msigdbr(species = "human", category = "H")
pathwaysH <- split(x = msigdbr_df_H$gene_symbol, f = msigdbr_df_H$gs_name)

msigdbr_df_C2 <- msigdbr(species = "human", category = "C2")
msigdbr_df_KEGG <- msigdbr_df_C2 %>% dplyr::filter(gs_subcat %in% "CP:KEGG")
msigdbr_df_REACTOME <- msigdbr_df_C2 %>% dplyr::filter(gs_subcat %in% "CP:REACTOME")

pathwaysC2_KEGG <- split(x = msigdbr_df_KEGG$gene_symbol, f = msigdbr_df_KEGG$gs_name)
pathwaysC2_REACTOME <- split(x = msigdbr_df_REACTOME$gene_symbol, f = msigdbr_df_REACTOME$gs_name)

ht <- function(d, m=5, n=m){
  # print the head and tail together
  rbind(HEAD = head(d,m), TAIL = tail(d,n))
}

library(readxl)    
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}



set.seed(444)
```

```{css, echo=FALSE}
.list-group-item.active, .list-group-item.active:hover, .list-group-item.active:focus {
    color: #ffffff;
    background-color: #333333;
}
p {
  color: #333333;
}
```

## Transcriptomics Analysis of Jurkat Cell Line Data.

This Rmarkdown file (adapted from Donagh Egan's script) records the result analysis of jurkat cell line transcriptomics data. Briefly, this dataset contain 4 conditions unstimulated, TCR stimulated, TCR+PD1 stimulated, and TCR+BTLA stimulated. For each of the stimulated condition, there are  `3 time points` which are **4 hours**, **20 hours**, and **48 hours** with three replicates each. So in total, there are 27 samples for stimulated cells of various conditions and 3 replicates for unstimulated. Below is a graphical representation of the experimental design:

```{r pressure, echo=FALSE, fig.cap="Experimental design", out.width = '100%',fig.align="center"}
knitr::include_graphics("./jurkat_experimental_design.png")

```

## Preprocessing

```{r Preprocessing}
## Formatting counts data ####
################################################################################

raw_counts <- readRDS( "./raw_counts.Rds")
raw_counts <- raw_counts[grepl("ENSG",raw_counts$ENSEMBL),]  ## select ENSG reads - human counts only 
raw_counts$ENSEMBL <- NULL ## remove ENSEMBL column
raw_counts <- aggregate(. ~SYMBOL, data=raw_counts, FUN=sum) ## sum over gene SYMBOL duplicates
raw_counts <- raw_counts %>% column_to_rownames("SYMBOL")
raw_counts <- raw_counts[!(row.names(raw_counts) %in% c("TNFRSF14","CD274","CD14")),] ## remove stimulatory human genes expressed in mouse cells
raw_counts <- raw_counts[,c(paste("S_", c(1:30), sep = ""))] ## order columns 

## Formatting metadata ####
################################################################################

exp_design <- readr::read_csv("./Novogene Nucleic Acid Sample Information Form_ Steinberger_2_2022.xlsx - Sample information form.csv",skip = 13)

exp_design <- exp_design[, c("*Sample Name (Tube Number)", "Remarks")]
names(exp_design)[1] <- "sample_number"
exp_design$sample_number <- paste("S_", exp_design$sample_number, sep = "")
exp_design <- exp_design %>% column_to_rownames("sample_number")
exp_design$Remarks <- gsub(" ", "_", exp_design$Remarks)
exp_design$Remarks <- as.factor(exp_design$Remarks)


exp_design <- exp_design %>%
  mutate(across('Remarks', str_replace, 'PDL1', 'PD1')) %>%
  mutate(across('Remarks', str_replace, 'HVEM', 'BTLA'))
## creating design matrix ####
################################################################################

condition <- exp_design$Remarks
designlimma <- as.data.frame(model.matrix(~ 0 + condition))
names(designlimma) <- gsub("condition", "", names(designlimma))

## normalizing counts using DESEQ2 ####
################################################################################

## DESEQ2 normalization
dds <- DESeqDataSetFromMatrix(countData = raw_counts,
                              colData = exp_design,
                              ~ Remarks)

keep <- rowSums(counts(dds)) >= 10 #QC
dds <- dds[keep,]

dds <- DESeq(dds) # Calculate DEGs

#Retrieven normalized data
dds <- estimateSizeFactors(dds) #Normalizatin procedure
normalized_counts <- counts(dds, normalized=TRUE)
normalized_counts <- log2(normalized_counts+1)

boxplot(normalized_counts,main="",outline=FALSE)
```


## Combining DEGs

```{r}
DEGs_all <- list()

DEGs_all[["StimvsUnstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "ctrl_4h", "unstimuliert_0"), 
                                                  format = "DataFrame"))
DEGs_all[["StimvsUnstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_20h", "unstimuliert_0"), 
                                                   format = "DataFrame"))
DEGs_all[["StimvsUnstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_48h", "unstimuliert_0"), 
                                                   format = "DataFrame"))
DEGs_all[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_all[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_all[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))
DEGs_all[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                          contrast = c("Remarks", "BTLA_4h", "ctrl_4h"),
                                                          format = "DataFrame"))
DEGs_all[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                           format = "DataFrame"))
DEGs_all[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                           format = "DataFrame"))

DEGs.long.all <- do.call(rbind, DEGs_all)
DEGs.long.all$gene <-sapply(strsplit(rownames(DEGs.long.all), "\\."), "[", 2)
DEGs.long.all$comparison <-sapply(strsplit(rownames(DEGs.long.all), "\\."), "[", 1)

DEGs.long.all <- DEGs.long.all %>% dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5))

saveRDS(DEGs.long.all, "Bulk_Signatures/Bulk_Signatures.rds")
```



## Figure 2: TCR Stimulation

```{r TCR-stim, warning=F}
##############################
### Process TCR Stimulated ###
##############################

DEGs <- list()

DEGs[["StimvsUnstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "ctrl_4h", "unstimuliert_0"), 
                                                  format = "DataFrame"))
DEGs[["StimvsUnstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_20h", "unstimuliert_0"), 
                                                   format = "DataFrame"))
DEGs[["StimvsUnstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_48h", "unstimuliert_0"), 
                                                   format = "DataFrame"))
DEGs.long <- do.call(rbind, DEGs)

DEGs.long$gene <-sapply(strsplit(rownames(DEGs.long), "\\."), "[", 2)
DEGs.long$comparison <-sapply(strsplit(rownames(DEGs.long), "\\."), "[", 1)
DEGs.wide <- reshape(DEGs.long[,c(2,6,7,8)], idvar = "gene", timevar = "comparison",direction = "wide")
colnames(DEGs.wide) <- gsub("log2FoldChange.","",colnames(DEGs.wide))
colnames(DEGs.wide) <- gsub("padj.","",colnames(DEGs.wide))

Merged.counts.stat <- normalized_counts
colnames(Merged.counts.stat) <- paste0(exp_design$Remarks,".",rownames(exp_design))
Merged.counts.stat <- merge(Merged.counts.stat,DEGs.wide, by.x = "row.names", by.y = "gene", all.x=T)
write.xlsx(Merged.counts.stat, rowNames = F, 
           file = "./Manuscript_Figures/figure2/Statistics_Stim_vs_Unstim.xlsx")

################
### Bar Plot ###
################

title_bar <- c("Stim vs Unstim 4h","Stim vs Unstim 20h","Stim vs Unstim 48h")
DEGs.selection <- lapply(DEGs, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
for (i in 1:length(DEGs.selection)){
  DEGs.selection[[i]]$direction <- ifelse(DEGs.selection[[i]]$log2FoldChange > log2(1.5), 
                                          "upregulated", "downregulated")
  DEGs.selection[[i]] <- DEGs.selection[[i]] %>%
    group_by(direction) %>%
    dplyr::count()
  DEGs.selection[[i]]$group <- title_bar[i]
}
DEGs.selection <- do.call(rbind, DEGs.selection)
DEGs.selection$n <- ifelse(DEGs.selection$direction == "downregulated", DEGs.selection$n*-1,DEGs.selection$n)
DEGs.selection$direction <- factor(DEGs.selection$direction, 
                                   levels = c("upregulated", "downregulated"))
DEGs.selection$group <- factor(DEGs.selection$group, 
                                   levels = title_bar)
# plot
library(hrbrthemes)

svg(paste("./Manuscript_Figures/figure2/DEG_CountStim_Unstim.svg", sep=""), 
    width = 16, height = 10)
ggplot(DEGs.selection, aes(fill=direction, y=n, x=group)) + 
  geom_bar(position="stack", stat="identity") + 
  geom_text(aes(label= ifelse(DEGs.selection$n > 0, n, -n), 
              x=group, 
              y=n),
          vjust = ifelse(DEGs.selection$n > 0, 2.5, -3.5) , size= 13, color = "white") +
  theme_ipsum() +
  theme(axis.text.y=element_blank(),
      axis.ticks.y=element_blank(), legend.title = element_blank(), 
      legend.text = element_text(size = 45, face = "bold"),
      axis.title.x  = element_blank(), 
      axis.text.x = element_text(size = 32, face = "bold"), axis.title.y  = element_blank(),
      )
dev.off()

#####################
### Venn diagrams ###
#####################

DEGs.selection <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.selection <- lapply(DEGs.selection , function(df) {rownames(df)})
DEGs.selection2 <- DEGs.selection
names(DEGs.selection2) <- c("4 hours", 
                            "20 hours",
                            "48 hours")
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
        x = DEGs.selection2,
        filename = "./Manuscript_Figures/figure2/Venn_Stim_vs_Unstim_upregulated.png",
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 2300 , 
        width = 2300 , 
        resolution = 400,
        disable.logging = T,
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = 2.5,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 2.5,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)

#Downregulated
DEGs.selection <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.selection <- lapply(DEGs.selection , function(df) {rownames(df)})
DEGs.selection2 <- DEGs.selection
names(DEGs.selection2) <- c("4 hours", 
                            "20 hours",
                            "48 hours")
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
        x = DEGs.selection2,
        filename = "./Manuscript_Figures/figure2/Venn_Stim_vs_Unstim_downregulated.png",
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 2300 , 
        width = 2300 , 
        resolution = 400,
        disable.logging = T,
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = 2.5,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 2.5,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)

####################
### Volcano plots ###
#####################

library(EnhancedVolcano)
DEGs.selection2 <- DEGs.long[,c(2,4,6,7,8)]
DEGs.selection2 <- DEGs.selection2[DEGs.selection2$gene != "",]
DEGs.selection2 <- DEGs.selection2[!(is.na(DEGs.selection2$padj) | DEGs.selection2$log2FoldChange==""), ]
colnames(DEGs.selection2) <- c("logFC","stat","adj.P.Val","GENE_SYMBOL","type")
#DEG2$type <-sub("-.*", "", DEG2$type) # it gives us only first part
g1 <- unique(DEGs.selection2$type)

title_volcano <- c("4h","20h","48h")

for (i in 1:length(g1)) {
  #i <- 30 
  DEGvolcano <- subset(DEGs.selection2, DEGs.selection2$type == g1[i])
  svg(paste("./Manuscript_Figures/figure2/VolcanoPlotStim_Unstim_",g1[i],".svg", sep=""), 
      width = 10, height = 10) #bg = 'transparent'
  p <-    EnhancedVolcano(DEGvolcano,
                          lab = DEGvolcano$GENE_SYMBOL,
                          x = "logFC",
                          y = "adj.P.Val",
                          ylab = "-Log10 adjusted p-val",
                          pCutoff = 0.05,
                          FCcutoff = log2(1.5),
                          col=c("black", "black", "black", "red3"),
                          colAlpha = 1,
                          title = paste0("Unstimulated vs Stimulated ",title_volcano[i]),
                          subtitle = "",
                          legendPosition = 'right',
                          legendLabSize=-1, legendIconSize=-1,titleLabSize = 35
                          #legend=c("","","",""), 
                          #,DrawConnectors = TRUE,
                          #widthConnectors = 0.2,
                          #colConnectors = 'grey30'
  )
  print(p)
  dev.off()
  print(p)
}

# upregulated
DEGs.selection.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)
DEGs.selection.up <- Reduce(intersect,DEGs.selection.up) %>% unique(.)
# DEGs.selection.up <- Reduce(c,DEGs.selection.up) %>% unique(.)
normalized_counts_TCR_up <- normalized_counts[DEGs.selection.up,TCR_design]

svg(paste("./Manuscript_Figures/figure2/HeatmapAllDEGSUpregulatedStim_Unstim.svg", sep=""), 
    width = 6, height = 6 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_TCR_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, name = " ",
                        column_title = paste0("Genes upregulated upon TCR-complex\n stimulation at all time points (", nrow(normalized_counts_TCR_up)," genes)" ), 
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"))
dev.off()

# downregulated
DEGs.selection.down <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.selection.down <- lapply(DEGs.selection.down, rownames)
DEGs.selection.down <- Reduce(intersect,DEGs.selection.down) %>% unique(.)
# DEGs.selection.down <- Reduce(c,DEGs.selection.down) %>% unique(.)
normalized_counts_TCR_down <- normalized_counts[DEGs.selection.down,TCR_design]

svg(paste("./Manuscript_Figures/figure2/HeatmapAllDEGSDownregulatedStim_Unstim.svg", sep=""), 
    width = 6, height = 6 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_TCR_down %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, name = " ",
                        column_title = paste0("Genes downregulated upon TCR-complex\n stimulation at all time points (",nrow(normalized_counts_TCR_down)," genes)"), 
                        column_title_gp = gpar(fontsize = 15, fontface = "bold") )
dev.off()


###########
## FGSEA ##
###########
DEGs_stat <- NULL
for(i in 1:length(DEGs)){
  DEGs_stat[[i]] <- DEGs[[i]]$stat
  names(DEGs_stat[[i]]) <- rownames(DEGs[[i]])
  DEGs_stat[[i]] <- sort(DEGs_stat[[i]], decreasing = T)
}
#Hallmark
fgseaRes_Hallmark <- NULL
for(i in 1:length(DEGs_stat)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat[[i]], minSize = 15)
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(padj < 0.05)
}
pathways_hallmark <- Reduce(c, list(fgseaRes_Hallmark[[1]]$pathway, 
                                    fgseaRes_Hallmark[[2]]$pathway, 
                                    fgseaRes_Hallmark[[3]]$pathway)) %>% 
  unique(.)
fgseaRes_Hallmark <- vector(mode='list', length=3)
names(fgseaRes_Hallmark) <- c("TCR 4h", "TCR 20h", "TCR 48h")
for(i in 1:length(DEGs_stat)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat[[i]], minSize = 15)
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(pathway %in% pathways_hallmark)
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::select(pathway, padj, NES)
  fgseaRes_Hallmark[[i]]$group <- names(fgseaRes_Hallmark)[i]
}
fgseaRes_Hallmark_plot <- do.call(rbind, fgseaRes_Hallmark)
fgseaRes_Hallmark_plot$group <- factor(fgseaRes_Hallmark_plot$group, 
                                   levels = c("TCR 4h", "TCR 20h", "TCR 48h"))
fgseaRes_Hallmark_plot$pathway <- gsub("HALLMARK_", "", fgseaRes_Hallmark_plot$pathway)
fgseaRes_Hallmark_plot$pathway <- gsub("_", " ", fgseaRes_Hallmark_plot$pathway)

pathway_order <- fgseaRes_Hallmark_plot %>% dplyr::filter(group %in% "TCR 20h") %>%
  arrange(NES) %>%
  pull(pathway)
fgseaRes_Hallmark_plot$pathway <- factor(fgseaRes_Hallmark_plot$pathway, 
                                   levels = pathway_order )
fgseaRes_Hallmark_plot$padj <- ifelse(fgseaRes_Hallmark_plot$padj > 0.05 , NA, fgseaRes_Hallmark_plot$padj)
fgseaRes_Hallmark_plot <- fgseaRes_Hallmark_plot %>%
  group_by(pathway) %>%
  filter(!all(is.na(padj))) %>%
  ungroup()


hallmark_bubble_plot <- ggplot(fgseaRes_Hallmark_plot, aes(x = group, y = pathway)) + 
  geom_point(aes(fill = NES, size = -log10(padj)), shape = 21) + 
  labs( x= "", y = "", fill = "NES", size = "-log10(padj)")  + 
  scale_fill_gradient2(low="blue", mid="white", high="red", oob = scales::squish) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", 
        axis.text = element_text(face = "bold", size = 16), 
        legend.text = element_text(face = "bold", size = 16), 
        legend.title = element_text(face = "bold", size = 16),
        title = element_text(face = "bold", size = 22) 
        )+
  labs(title = "Gene Set \nEnrichment Analysis")

svg(paste("./Manuscript_Figures/figure2/HallmarkStim_Unstim_minsize15.svg", sep=""), 
    width = 10, height = 10 )
hallmark_bubble_plot
dev.off()




####################
## Motif Analysis ##
####################
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

DEGs_signif <- DEGs
for (i in 1:length(DEGs_signif)){DEGs_signif[[i]] <- DEGs[[i]] %>% dplyr::filter(padj < 0.05)}

# Upregulated
DEGs_signif_up <- DEGs_signif
for (i in 1:length(DEGs_signif_up)){
  DEGs_signif_up[[i]] <- DEGs_signif_up[[i]] %>% 
    dplyr::filter(log2FoldChange > log2(1.5)) %>% rownames(.)
}

motif_TCR_up <- as.list(rep(NA, length(DEGs_signif_up)))
for(i in 1:length(motif_TCR_up)){
    tryCatch(
      expr = {
        temp <- intersect(DEGs_signif_up[[i]], colnames(motifRankings))
        motif_TCR_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                         motifAnnot=motifAnnotations_hgnc)
      },
      error = function(e){
        message('insufficent genes')
        motif_TCR_up[[i]] <- NULL
      })
}
names(motif_TCR_up) <- c("TCR 4h", "TCR 20h", "TCR 48h")
for (i in 1:length(motif_TCR_up)){
  motif_TCR_up[[i]] <- motif_TCR_up[[i]] %>%
    filter(str_detect(TF_highConf, "directAnnotation")) %>%
    dplyr::select(TF_highConf, NES)
  motif_TCR_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- sub(" .*", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub(";", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$group <- names(motif_TCR_up)[i]
}

write.xlsx(motif_TCR_up, "Manuscript_Figures/others/motif_TCR.xlsx")


motif_TCR_up_combined <- do.call(rbind, motif_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% unique(.)

#selection
to_select_TCR_up <- motif_TCR_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_TCR_up_combined <- motif_TCR_up_combined %>% dplyr::filter(TF_highConf %in% to_select_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_TCR_up <- (max(motif_TCR_up_combined$NES) + min(motif_TCR_up_combined$NES))/ 2

# Ordering plot
motif_TCR_up_combined$group <- factor(motif_TCR_up_combined$group, 
                                   levels = c("TCR 4h", "TCR 20h", "TCR 48h"))
motif_TCR_up_combined <- motif_TCR_up_combined[order(motif_TCR_up_combined$group,
                                                     motif_TCR_up_combined$NES),]
TF_order_TCR_up <- motif_TCR_up_combined$TF_highConf %>% unique(.)
motif_TCR_up_combined$TF_highConf <- factor(motif_TCR_up_combined$TF_highConf, 
                                   levels = TF_order_TCR_up)
#plot
motif_TCR_up_plot <- ggplot(motif_TCR_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_TCR_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold"),
        plot.title = element_text(size = 27, face = "bold")) +
  labs(title = "High confidence transcription factor for genes upregulated upon TCR-complex stimulation")
svg(paste("./Manuscript_Figures/figure2/MotifUpregulatedStim_Unstim_NES2.svg", sep=""), 
    width = 24, height = 3 )
motif_TCR_up_plot
dev.off()


# Downregulated
DEGs_signif_down <- DEGs_signif
for (i in 1:length(DEGs_signif_down)){
  DEGs_signif_down[[i]] <- DEGs_signif_down[[i]] %>% 
    dplyr::filter(log2FoldChange < -log2(1.5)) %>% rownames(.)
}

motif_TCR_down <- as.list(rep(NA, length(DEGs_signif_down)))
for(i in 1:length(motif_TCR_down)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_down[[i]], colnames(motifRankings))
      motif_TCR_down[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                     motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_TCR_down[[i]] <- NULL
    })
}
names(motif_TCR_down) <- c("TCR 4h", "TCR 20h", "TCR 48h")
for (i in 1:length(motif_TCR_down)){
  motif_TCR_down[[i]] <- motif_TCR_down[[i]] %>%
    filter(str_detect(TF_highConf, "directAnnotation")) %>%
    dplyr::select(TF_highConf, NES)
  motif_TCR_down[[i]]$TF_highConf <-  gsub('\\.', "", motif_TCR_down[[i]]$TF_highConf)
  motif_TCR_down[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_TCR_down[[i]]$TF_highConf)
  motif_TCR_down[[i]]$TF_highConf <- sub(" .*", "", motif_TCR_down[[i]]$TF_highConf)
  motif_TCR_down[[i]]$TF_highConf <- gsub(";", "", motif_TCR_down[[i]]$TF_highConf)
  motif_TCR_down[[i]]$group <- names(motif_TCR_down)[i]
}


motif_TCR_down_combined <- do.call(rbind, motif_TCR_down)
motif_TCR_down_combined <- motif_TCR_down_combined %>% unique(.)

#selection
to_select_TCR_down <- motif_TCR_down_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_TCR_down_combined <- motif_TCR_down_combined %>% dplyr::filter(TF_highConf %in% to_select_TCR_down)
motif_TCR_down_combined <- motif_TCR_down_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_TCR_down <- (max(motif_TCR_down_combined$NES) + min(motif_TCR_down_combined$NES))/ 2

# Ordering plot
motif_TCR_down_combined$group <- factor(motif_TCR_down_combined$group, 
                                      levels = c("TCR 4h", "TCR 20h", "TCR 48h"))
motif_TCR_down_combined <- motif_TCR_down_combined[order(motif_TCR_down_combined$group,
                                                     motif_TCR_down_combined$NES),]
TF_order_TCR_down <- motif_TCR_down_combined$TF_highConf %>% unique(.)
motif_TCR_down_combined$TF_highConf <- factor(motif_TCR_down_combined$TF_highConf, 
                                            levels = TF_order_TCR_down)
#plot
motif_TCR_down_plot <- ggplot(motif_TCR_down_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_TCR_down) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold"),
        plot.title = element_text(size = 27, face = "bold")) +
  labs(title = "High confidence transcription factor for genes downregulated upon TCR-complex stimulation")
svg(paste("./Manuscript_Figures/figure2/MotifDownregulatedStim_Unstim.svg", sep=""), 
    width = 24, height = 3 )
motif_TCR_down_plot
dev.off()



################################
## Kinase Enrichment Analysis ##
################################
library(httr)
library(jsonlite)

DEGs.selection.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
  arrange(padj) %>%
  head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]

  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_TCR.xlsx")


# KEA visualization
top_10 <- lapply(KEA_results, function(i) head(i, 10)) 
for (i in 1:length(top_10)){top_10[[i]] <- top_10[[i]] %>% pull(TF)}
top_10 <- Reduce(c,top_10) %>% unique(.)
KEA_results_subset <- KEA_results

type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% dplyr::filter(TF %in% top_10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)

pathway_order <- to_plot %>% dplyr::filter(group %in% "4h") %>%
  arrange(desc(logScore)) %>%
  pull(TF)
to_plot$TF <- factor(to_plot$TF, levels = pathway_order )
to_plot$logScore <- ifelse(to_plot$Score < 100, to_plot$logScore, NA)


KEA_bubble_plot <- ggplot(to_plot, aes(x = group, y = TF)) + 
  geom_point(aes(color = logScore ), size = 5) + 
  labs( x= "", y = "",size = "Log-score")  + 
  scale_color_gradient2(low="yellow", mid = "orange",high="red", midpoint = mean(to_plot$logScore, na.rm = T),na.value = "#FFFFFF" ) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", 
        axis.text = element_text(face = "bold", size = 16), 
        legend.text = element_text(face = "bold", size = 16), 
        legend.title = element_text(face = "bold", size = 16),
        title = element_text(face = "bold", size = 22) 
        )+
  labs(title = "Kinase Enrichment Analysis")

svg(paste("./Manuscript_Figures/figure2/KEA_TCR_Plot.svg", sep=""), 
    width = 10, height = 10 )
KEA_bubble_plot
dev.off()


################################
## Kinase Enrichment Analysis ##
################################
#TCR
DEGs.selection.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_TCR.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_TCR.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("4h",1), rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("4h", "20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("4h", "20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

ha1 = HeatmapAnnotation(type = type$group,name = " ",
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")), which = "row", annotation_label = " ")

svg(paste0("./Manuscript_Figures/figure2/KEA_TCR_Heatmap.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun, left_annotation =  ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()
```


# Figure 3: PD1 and BTLA analyis

```{r}
library(PCAtools)

# PCA Plot
exp_design2 <- exp_design
exp_design2$type <- as.factor(gsub("\\_.*","",exp_design2$Remarks))
exp_design2$Group <- as.factor(gsub(".*_", "", exp_design2$Remarks))
exp_design2$type <- ifelse(exp_design2$type == "unstimuliert", "unstimulated", 
                           as.character(exp_design2$type))
exp_design2$Group <- factor(exp_design2$Group , levels = c("0","4h","20h", "48h"))
exp_design2$type <- factor(exp_design2$type , levels = c("unstimulated","ctrl",
                                                         "PD1", "BTLA"))
# PCA Plot
p <- PCAtools::pca(normalized_counts, metadata = exp_design2, removeVar = 0.1)


svg(paste("Manuscript_Figures/figure3/PCA.svg", sep=""), 
    width = 20, height = 8 )
print(biplot(p,
  lab = NULL,
  shape = 'type',
  colby = 'Group', 
  colkey = c("0" = "#F8766D", "4h" = "#7CAE00", "20h" = "#00BFC4", "48h" = "#C77CFF"),
  hline = c(-25, 0, 25), vline = c(-25, 0, 25),
  vlineType = c('dotdash', 'solid', 'dashed'),
  gridlines.major = FALSE, gridlines.minor = FALSE,
  pointSize = 6,
  titleLabSize = 25,
  legendPosition = 'left', legendLabSize = 14, legendIconSize = 8.0,
  drawConnectors = FALSE,
  title = 'PCA plot',
  subtitle = 'Stimulation condition and timepoints',
  # ellipse config
  ellipse = TRUE,
  ellipseType = 't',
  ellipseLevel = 0.95,
  ellipseFill = T,
  ellipseAlpha = 1/5,
  ellipseLineSize = 0
  )
)
dev.off()



###################
## Volcano plots ##
###################
plot_title <- c("4h", "20h", "48h")
# PD1
DEGs_PD1 <- list()

DEGs_PD1[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))
DEGs_PD1_long <- do.call(rbind, DEGs_PD1)

DEGs_PD1_long$gene <-sapply(strsplit(rownames(DEGs_PD1_long), "\\."), "[", 2)
DEGs_PD1_long$comparison <-sapply(strsplit(rownames(DEGs_PD1_long), "\\."), "[", 1)
DEGs_PD1_wide <- reshape(DEGs_PD1_long[,c(2,6,7,8)], idvar = "gene", timevar = "comparison",direction = "wide")
colnames(DEGs_PD1_wide) <- gsub("log2FoldChange.","",colnames(DEGs_PD1_wide))
colnames(DEGs_PD1_wide) <- gsub("padj.","",colnames(DEGs_PD1_wide))

Merged.counts.stat <- normalized_counts
colnames(Merged.counts.stat) <- paste0(exp_design$Remarks,".",rownames(exp_design))
Merged.counts.stat <- merge(Merged.counts.stat,DEGs_PD1_wide, by.x = "row.names", by.y = "gene", all.x=T)
# write.xlsx(Merged.counts.stat, rowNames = F, 
#            file = "./Manuscript_Figures/figure3/Statistics_PD1stim_vs_TCRstim.xlsx")

library(EnhancedVolcano)
DEGs.selection2_PD1 <- DEGs_PD1_long[,c(2,4,6,7,8)]
DEGs.selection2_PD1 <- DEGs.selection2_PD1[DEGs.selection2_PD1$gene != "",]
DEGs.selection2_PD1 <- DEGs.selection2_PD1[!(is.na(DEGs.selection2_PD1$padj) | 
                                                 DEGs.selection2_PD1$log2FoldChange==""), ]
colnames(DEGs.selection2_PD1) <- c("logFC","stat","adj.P.Val","GENE_SYMBOL","type")
#DEG2$type <-sub("-.*", "", DEG2$type) # it gives us only first part
g1_PD1 <- unique(DEGs.selection2_PD1$type)

for (i in 1:length(g1_PD1)) {
  #i <- 30 
  DEGvolcano <- subset(DEGs.selection2_PD1, DEGs.selection2_PD1$type == g1_PD1[i])
  svg(paste("./Manuscript_Figures/figure3/VolcanoPlotStim_Unstim_",g1_PD1[i],".svg", sep=""), 
      width = 7, height = 4.5 ) #bg = 'transparent'
  p <-    EnhancedVolcano(DEGvolcano,
                          lab = DEGvolcano$GENE_SYMBOL,
                          x = "logFC",
                          y = "adj.P.Val",
                          ylab = "-Log10 adjusted p-val",
                          pCutoff = 0.05,
                          FCcutoff = log2(1.5),
                          col=c("black", "black", "black", "red3"),
                          colAlpha = 1,
                          title = paste0("",plot_title[i]),
                          subtitle = "",
                          legendPosition = 'right',titleLabSize = 19,
                          legendLabSize=-1, legendIconSize=-1 #legend=c("","","",""), 
                          #,DrawConnectors = TRUE,
                          #widthConnectors = 0.2,
                          #colConnectors = 'grey30'
  )+ theme(plot.title = element_text(hjust = 0.5))
  print(p)
  dev.off()
}

# BTLA
DEGs_BTLA <- list()

DEGs_BTLA[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                            contrast = c("Remarks", "BTLA_4h", "ctrl_4h"), 
                                                            format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                             contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                             format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                             contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                             format = "DataFrame"))
DEGs_BTLA_long <- do.call(rbind, DEGs_BTLA)

DEGs_BTLA_long$gene <-sapply(strsplit(rownames(DEGs_BTLA_long), "\\."), "[", 2)
DEGs_BTLA_long$comparison <-sapply(strsplit(rownames(DEGs_BTLA_long), "\\."), "[", 1)
DEGs_BTLA_wide <- reshape(DEGs_BTLA_long[,c(2,6,7,8)], idvar = "gene", timevar = "comparison",direction = "wide")
colnames(DEGs_BTLA_wide) <- gsub("log2FoldChange.","",colnames(DEGs_BTLA_wide))
colnames(DEGs_BTLA_wide) <- gsub("padj.","",colnames(DEGs_BTLA_wide))

Merged.counts.stat <- normalized_counts
colnames(Merged.counts.stat) <- paste0(exp_design$Remarks,".",rownames(exp_design))
Merged.counts.stat <- merge(Merged.counts.stat,DEGs_BTLA_wide, by.x = "row.names", by.y = "gene", all.x=T)
write.xlsx(Merged.counts.stat, rowNames = F, 
           file = "./Manuscript_Figures/figure3/Statistics_BTLAstim_vs_TCRstim.xlsx")


DEGs.selection2_BTLA <- DEGs_BTLA_long[,c(2,4,6,7,8)]
DEGs.selection2_BTLA <- DEGs.selection2_BTLA[DEGs.selection2_BTLA$gene != "",]
DEGs.selection2_BTLA <- DEGs.selection2_BTLA[!(is.na(DEGs.selection2_BTLA$padj) | 
                                                 DEGs.selection2_BTLA$log2FoldChange==""), ]
colnames(DEGs.selection2_BTLA) <- c("logFC","stat","adj.P.Val","GENE_SYMBOL","type")
#DEG2$type <-sub("-.*", "", DEG2$type) # it gives us only first part
g1_BTLA <- unique(DEGs.selection2_BTLA$type)

for (i in 1:length(g1_BTLA)) {
  #i <- 30 
  DEGvolcano <- subset(DEGs.selection2_BTLA, DEGs.selection2_BTLA$type == g1_BTLA[i])
  svg(paste("./Manuscript_Figures/figure3/VolcanoPlotStim_Unstim_",g1_BTLA[i],".svg", sep=""), 
      width = 7, height = 4.5 ) #bg = 'transparent'
  p <-    EnhancedVolcano(DEGvolcano,
                          lab = DEGvolcano$GENE_SYMBOL,
                          x = "logFC",
                          y = "adj.P.Val",
                          ylab = "-Log10 adjusted p-val",
                          pCutoff = 0.05,
                          FCcutoff = log2(1.5),
                          col=c("black", "black", "black", "red3"),
                          colAlpha = 1,
                          title = paste0("",plot_title[i]),
                          subtitle = "",
                          legendPosition = 'right',titleLabSize = 19,
                          legendLabSize=-1, legendIconSize=-1 #legend=c("","","",""), 
                          #,DrawConnectors = TRUE,
                          #widthConnectors = 0.2,
                          #colConnectors = 'grey30'
  ) + theme(plot.title = element_text(hjust = 0.5))
  print(p)
  dev.off()
}


#######################################
#### Heatmap of PD1 and BTLA DEGs ####
#######################################

## 20 hours
library(ComplexHeatmap)
pat <- "(_20h)"
stim_design_20h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("TCR 20h",3), rep("PD1 20h",3), 
                             rep("BTLA 20h",3) ))
type$group <- factor(type$group, levels = c("TCR 20h", "PD1 20h", "BTLA 20h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("TCR 20h" = "#cccccc", "PD1 20h" = "#00aa00", "BTLA 20h" = "#c837ab")))

library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)


# upregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.up.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.PD1 <- lapply(DEGs.selection.up.PD1, rownames)

DEGs.selection.up.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.BTLA <- lapply(DEGs.selection.up.BTLA, rownames)

DEGs.selection.up.20h <- unique(c(DEGs.selection.up.PD1$PD1StimvsTCRstim_20, 
                                  DEGs.selection.up.BTLA$BTLAStimvsTCRstim_20))
DEGs.selection.up.20h <- intersect(DEGs.selection.up.20h, DEGs.TCR.up$StimvsUnstim_20)

normalized_counts_20h_up <- normalized_counts[DEGs.selection.up.20h, stim_design_20h]
normalized_counts_20h_up <- normalized_counts_20h_up[rowSums(
  !as.matrix(normalized_counts_20h_up)) < ncol(normalized_counts_20h_up), ]

svg(paste("./Manuscript_Figures/figure3/Heatmap_upregulated_20h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_20h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes upregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 20 hour", name = " ")
dev.off()

# Venn with p-val
PD1_signif <- DEGs_PD1$PD1StimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif <- DEGs_BTLA$BTLAStimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()

to_plot <- list(
  "PD1" = intersect(PD1_signif, DEGs.selection.up.20h),
  "BTLA" = intersect(BTLA_signif, DEGs.selection.up.20h)
)
myCol <- brewer.pal(2, "Pastel2")
myCol <- myCol[1:2]
venn.diagram(
        x = to_plot,
        filename = "./Manuscript_Figures/figure3/Venn_20h_Up_TCR.png",
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 2300 , 
        width = 2300 , 
        resolution = 300,
        disable.logging = T,
        main = "20 hours Upregulated TCR",
        main.cex = 3,
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = 1.7,
        fontface = "bold",
        fontfamily = "sans"
)

# downregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.down.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.PD1 <- lapply(DEGs.selection.down.PD1, rownames)

DEGs.selection.down.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.BTLA <- lapply(DEGs.selection.down.BTLA, rownames)

DEGs.selection.down.20h <- unique(c(DEGs.selection.down.PD1$PD1StimvsTCRstim_20, 
                                    DEGs.selection.down.BTLA$BTLAStimvsTCRstim_20))
DEGs.selection.down.20h <- intersect(DEGs.selection.down.20h, DEGs.TCR.up$StimvsUnstim_20)

normalized_counts_20h_up <- normalized_counts[DEGs.selection.down.20h, stim_design_20h]
normalized_counts_20h_up <- normalized_counts_20h_up[rowSums(
  !as.matrix(normalized_counts_20h_up)) < ncol(normalized_counts_20h_up), ]


svg(paste("./Manuscript_Figures/figure3/Heatmap_downregulated_20h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_20h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes downregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 20 hour", name = " ")
dev.off()


# Venn with p-val
PD1_signif <- DEGs_PD1$PD1StimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif <- DEGs_BTLA$BTLAStimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()

to_plot <- list(
  "PD1" = intersect(PD1_signif, DEGs.selection.down.20h),
  "BTLA" = intersect(BTLA_signif, DEGs.selection.down.20h)
)
myCol <- brewer.pal(2, "Pastel2")
myCol <- myCol[1:2]
venn.diagram(
        x = to_plot,
        filename = "./Manuscript_Figures/figure3/Venn_20h_Down_TCR.svg",
        output=TRUE,
        
        # Output features
        imagetype="svg" ,
        height = 2300 , 
        width = 2300 , 
        resolution = 300,
        disable.logging = T,
        main = "20 hours Downregulated TCR",
        main.cex = 3,
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = 1.7,
        fontface = "bold",
        fontfamily = "sans"
)


## 48 hours
library(ComplexHeatmap)
pat <- "(_48h)"
stim_design_48h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("TCR 48h",3), rep("PD1 48h",3), 
                             rep("BTLA 48h",3) ))
type$group <- factor(type$group, levels = c("TCR 48h", "PD1 48h", "BTLA 48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("TCR 48h" = "#cccccc", "PD1 48h" = "#00aa00", "BTLA 48h" = "#c837ab")))

library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)


# upregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.up.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.PD1 <- lapply(DEGs.selection.up.PD1, rownames)

DEGs.selection.up.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.BTLA <- lapply(DEGs.selection.up.BTLA, rownames)

DEGs.selection.up.48h <- unique(c(DEGs.selection.up.PD1$PD1StimvsTCRstim_48, 
                                  DEGs.selection.up.BTLA$BTLAStimvsTCRstim_48))
DEGs.selection.up.48h <- intersect(DEGs.selection.up.48h, DEGs.TCR.up$StimvsUnstim_48)

normalized_counts_48h_up <- normalized_counts[DEGs.selection.up.48h, stim_design_48h]
normalized_counts_48h_up <- normalized_counts_48h_up[rowSums(
  !as.matrix(normalized_counts_48h_up)) < ncol(normalized_counts_48h_up), ]



svg(paste("./Manuscript_Figures/figure3/Heatmap_upregulated_48h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_48h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes upregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 48 hour", name = " ")
dev.off()


# Venn with pval
PD1_signif <- DEGs_PD1$PD1StimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif <- DEGs_BTLA$BTLAStimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()


to_plot <- list(
  "PD1" = intersect(PD1_signif, DEGs.selection.up.48h),
  "BTLA" = intersect(BTLA_signif, DEGs.selection.up.48h)
)
myCol <- brewer.pal(2, "Pastel2")
myCol <- myCol[1:2]
venn.diagram(
  x = to_plot,
  filename = "./Manuscript_Figures/figure3/Venn_48h_Up_TCR.png",
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 300,
  disable.logging = T,
  main = "48 hours Downregulated TCR",
  main.cex = 3,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = 1.7,
  fontface = "bold",
  fontfamily = "sans"
)

# downregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.down.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.PD1 <- lapply(DEGs.selection.down.PD1, rownames)

DEGs.selection.down.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.BTLA <- lapply(DEGs.selection.down.BTLA, rownames)

DEGs.selection.down.48h <- unique(c(DEGs.selection.down.PD1$PD1StimvsTCRstim_48, 
                                  DEGs.selection.down.BTLA$BTLAStimvsTCRstim_48))
DEGs.selection.down.48h <- intersect(DEGs.selection.down.48h, DEGs.TCR.up$StimvsUnstim_48)

normalized_counts_48h_up <- normalized_counts[DEGs.selection.down.48h, stim_design_48h]
normalized_counts_48h_up <- normalized_counts_48h_up[rowSums(
  !as.matrix(normalized_counts_48h_up)) < ncol(normalized_counts_48h_up), ]


svg(paste("./Manuscript_Figures/figure3/Heatmap_downregulated_48h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_48h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes downregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 48 hour", name = " ")
dev.off()



# Venn with Pval
PD1_signif <- DEGs_PD1$PD1StimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif <- DEGs_BTLA$BTLAStimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()


to_plot <- list(
  "PD1" = intersect(PD1_signif, DEGs.selection.down.48h),
  "BTLA" = intersect(BTLA_signif, DEGs.selection.down.48h)
)
myCol <- brewer.pal(2, "Pastel2")
myCol <- myCol[1:2]
venn.diagram(
  x = to_plot,
  filename = "./Manuscript_Figures/figure3/Venn_48h_Down_TCR.png",
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 300,
  disable.logging = T,
  main = "48 hours Downregulated TCR",
  main.cex = 3,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = 1.7,
  fontface = "bold",
  fontfamily = "sans"
)


########################
### Venn Diagram All ###
########################

# Upregulated
DEGs.TCR.all <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.TCR.all <- lapply(DEGs.TCR.all, rownames)
DEGs.TCR.all <- c(DEGs.TCR.all$StimvsUnstim_20, DEGs.TCR.all$StimvsUnstim_48) %>% unique(.)

PD1_signif_20 <- DEGs_PD1$PD1StimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif_20 <- DEGs_BTLA$BTLAStimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
PD1_signif_48 <- DEGs_PD1$PD1StimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif_48 <- DEGs_BTLA$BTLAStimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()


to_plot <- list(
  "PD1 20h" = intersect(PD1_signif_20, DEGs.TCR.all),
  "BTLA 20h" = intersect(BTLA_signif_20, DEGs.TCR.all),
  "PD1 48h" = intersect(PD1_signif_48, DEGs.TCR.all),
  "BTLA 48h" = intersect(BTLA_signif_48, DEGs.TCR.all)
)
myCol <- brewer.pal(4, "Pastel2")
myCol <- myCol[1:4]
venn.diagram(
  x = to_plot,
  filename = "./Manuscript_Figures/figure3/Venn_PD1_BTLA_Upregulated.png",
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 300,
  disable.logging = T,
  main = "Upregulated",
  main.cex = 3,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = 1.7,
  fontface = "bold",
  fontfamily = "sans"
)

combos <- Reduce(c,lapply(2:length(to_plot), 
            function(x) combn(1:length(to_plot),x,simplify=FALSE) ))

test <- lapply(combos, function(x) Reduce(intersect,to_plot[x]) )

map_name = setNames(c(1,2,3,4), c("PD1_20", "BTLA_20", "PD1_48", "BTLA_48"))
for (i in 1:length(combos)){
  combos[[i]] <- names(map_name[unlist(combos[[i]])])
  combos[[i]] <- paste(combos[[i]], collapse = '.')
  names(combos)[i] <- combos[[i]]
}
names(test) <- names(combos)
write.xlsx(test, "Manuscript_Figures/figure3/venn_genes_upregulated_TCR_attenuated.xlsx")


# Downregulated
DEGs.TCR.all <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.TCR.all <- lapply(DEGs.TCR.all, rownames)
DEGs.TCR.all <- c(DEGs.TCR.all$StimvsUnstim_20, DEGs.TCR.all$StimvsUnstim_48) %>% unique(.)


PD1_signif_20 <- DEGs_PD1$PD1StimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif_20 <- DEGs_BTLA$BTLAStimvsTCRstim_20 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
PD1_signif_48 <- DEGs_PD1$PD1StimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()
BTLA_signif_48 <- DEGs_BTLA$BTLAStimvsTCRstim_48 %>% 
  dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5)) %>% 
  row.names()


to_plot <- list(
  "PD1 20h" = intersect(PD1_signif_20, DEGs.TCR.all),
  "BTLA 20h" = intersect(BTLA_signif_20, DEGs.TCR.all),
  "PD1 48h" = intersect(PD1_signif_48, DEGs.TCR.all),
  "BTLA 48h" = intersect(BTLA_signif_48, DEGs.TCR.all)
)
myCol <- brewer.pal(4, "Pastel2")
myCol <- myCol[1:4]
venn.diagram(
  x = to_plot,
  filename = "./Manuscript_Figures/figure3/Venn_PD1_BTLA_Downregulated.png",
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 300,
  disable.logging = T,
  main = "Downregulated",
  main.cex = 3,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = 1.7,
  fontface = "bold",
  fontfamily = "sans"
)
combos <- Reduce(c,lapply(2:length(to_plot), 
            function(x) combn(1:length(to_plot),x,simplify=FALSE) ))

test <- lapply(combos, function(x) Reduce(intersect,to_plot[x]) )

map_name = setNames(c(1,2,3,4), c("PD1_20", "BTLA_20", "PD1_48", "BTLA_48"))
for (i in 1:length(combos)){
  combos[[i]] <- names(map_name[unlist(combos[[i]])])
  combos[[i]] <- paste(combos[[i]], collapse = '.')
  names(combos)[i] <- combos[[i]]
}
names(test) <- names(combos)
write.xlsx(test, "Manuscript_Figures/figure3/venn_genes_downregulated_TCR_attenuated.xlsx")



######################
## Pathway analysis ##
######################
# PD1
DEGs_stat_PD1 <- NULL
for(i in 1:length(DEGs_PD1)){
  DEGs_stat_PD1[[i]] <- DEGs_PD1[[i]]$stat
  names(DEGs_stat_PD1[[i]]) <- rownames(DEGs_PD1[[i]])
  DEGs_stat_PD1[[i]] <- sort(DEGs_stat_PD1[[i]], decreasing = T)
}
#Hallmark
fgseaRes_Hallmark <- NULL
for(i in 1:length(DEGs_stat_PD1)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_PD1[[i]])
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(padj < 0.05)
}
pathways_hallmark <- Reduce(c, list(fgseaRes_Hallmark[[1]]$pathway, fgseaRes_Hallmark[[2]]$pathway, fgseaRes_Hallmark[[3]]$pathway)) %>% 
  unique(.)
fgseaRes_Hallmark <- vector(mode='list', length=3)
names(fgseaRes_Hallmark) <- c("PD1 4h", "PD1 20h", "PD1 48h")
for(i in 1:length(DEGs_stat_PD1)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_PD1[[i]])
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(pathway %in% pathways_hallmark)
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::select(pathway, padj, NES)
  fgseaRes_Hallmark[[i]]$group <- names(fgseaRes_Hallmark)[i]
}
fgseaRes_Hallmark_plot <- do.call(rbind, fgseaRes_Hallmark)
fgseaRes_Hallmark_plot$group <- factor(fgseaRes_Hallmark_plot$group, 
                                   levels = c("PD1 4h", "PD1 20h", "PD1 48h"))
fgseaRes_Hallmark_plot$pathway <- gsub("HALLMARK_", "", fgseaRes_Hallmark_plot$pathway)
fgseaRes_Hallmark_plot$pathway <- gsub("_", " ", fgseaRes_Hallmark_plot$pathway)

pathway_order <- fgseaRes_Hallmark_plot %>% dplyr::filter(group %in% "PD1 20h") %>%
  arrange(NES) %>%
  pull(pathway)
fgseaRes_Hallmark_plot$pathway <- factor(fgseaRes_Hallmark_plot$pathway, 
                                   levels = pathway_order )
fgseaRes_Hallmark_plot$padj <- ifelse(fgseaRes_Hallmark_plot$padj > 0.05 , NA, fgseaRes_Hallmark_plot$padj)
fgseaRes_Hallmark_plot <- fgseaRes_Hallmark_plot %>%
  group_by(pathway) %>%
  filter(!all(is.na(padj))) %>%
  ungroup()

hallmark_bubble_plot <- ggplot(fgseaRes_Hallmark_plot, aes(x = group, y = pathway)) + 
  geom_point(aes(fill = NES, size = -log10(padj)), shape = 21) + 
  labs( x= "", y = "", fill = "NES", size = "-log10(padj)")  + 
  scale_fill_gradient2(low="blue", mid="white", high="red", oob = scales::squish) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", axis.text = element_text(face = "bold", size = 12), 
        legend.text = element_text(face = "bold", size = 10), 
        legend.title = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14) 
        )+
  labs(title = "Gene Set Enrichment Analysis (PD1)")

svg(paste("./Manuscript_Figures/figure3/HallmarkPD1_Stim.svg", sep=""), 
    width = 9, height = 10 )
hallmark_bubble_plot
dev.off()



library(enrichplot)
for (i in 1:length(DEGs_PD1)){
  test <- DEGs_PD1[[i]] %>%
    dplyr::mutate(gene = rownames(.))%>%
    pull(stat, gene)
  
  
  svg(paste0("./Manuscript_Figures/supp/Hallmark_G2M", names(DEGs_PD1)[i],".svg" ), 
    width = 9, height = 10 )
  print(plotEnrichment(pathwaysH[["HALLMARK_G2M_CHECKPOINT"]],
                 test) + labs(title= paste0("G2M CHECKPOINT_", names(DEGs_PD1)[i]) ))
  dev.off()
}

for (i in 1:length(DEGs_PD1)){
  test <- DEGs_PD1[[i]] %>%
  dplyr::mutate(gene = rownames(.))%>%
  pull(stat, gene)
  
  
  svg(paste0("./Manuscript_Figures/supp/Hallmark_NFKB", names(DEGs_PD1)[i],".svg" ), 
    width = 9, height = 10 )
  print(plotEnrichment(pathwaysH[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]],
                 test) + labs(title= paste0("TNFA SIGNALING VIA NFKB_", names(DEGs_PD1)[i]) ))
  dev.off()
}


# BTLA
DEGs_stat_BTLA <- NULL
for(i in 1:length(DEGs_BTLA)){
  DEGs_stat_BTLA[[i]] <- DEGs_BTLA[[i]]$stat
  names(DEGs_stat_BTLA[[i]]) <- rownames(DEGs_BTLA[[i]])
  DEGs_stat_BTLA[[i]] <- sort(DEGs_stat_BTLA[[i]], decreasing = T)
}
#Hallmark
fgseaRes_Hallmark <- NULL
for(i in 1:length(DEGs_stat_BTLA)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_BTLA[[i]])
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(padj < 0.05)
}
pathways_hallmark <- Reduce(c, list(fgseaRes_Hallmark[[1]]$pathway, fgseaRes_Hallmark[[2]]$pathway, fgseaRes_Hallmark[[3]]$pathway)) %>% 
  unique(.)
fgseaRes_Hallmark <- vector(mode='list', length=3)
names(fgseaRes_Hallmark) <- c("BTLA 4h", "BTLA 20h", "BTLA 48h")
for(i in 1:length(DEGs_stat_BTLA)){
  fgseaRes_Hallmark[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_BTLA[[i]])
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::filter(pathway %in% pathways_hallmark)
  fgseaRes_Hallmark[[i]] <- fgseaRes_Hallmark[[i]] %>% dplyr::select(pathway, padj, NES)
  fgseaRes_Hallmark[[i]]$group <- names(fgseaRes_Hallmark)[i]
}
fgseaRes_Hallmark_plot <- do.call(rbind, fgseaRes_Hallmark)
fgseaRes_Hallmark_plot$group <- factor(fgseaRes_Hallmark_plot$group, 
                                       levels = c("BTLA 4h", "BTLA 20h", "BTLA 48h"))
fgseaRes_Hallmark_plot$pathway <- gsub("HALLMARK_", "", fgseaRes_Hallmark_plot$pathway)
fgseaRes_Hallmark_plot$pathway <- gsub("_", " ", fgseaRes_Hallmark_plot$pathway)

pathway_order <- fgseaRes_Hallmark_plot %>% dplyr::filter(group %in% "BTLA 20h") %>%
  arrange(NES) %>%
  pull(pathway)
fgseaRes_Hallmark_plot$pathway <- factor(fgseaRes_Hallmark_plot$pathway, 
                                         levels = pathway_order )
fgseaRes_Hallmark_plot$padj <- ifelse(fgseaRes_Hallmark_plot$padj > 0.05 , NA, fgseaRes_Hallmark_plot$padj)
fgseaRes_Hallmark_plot <- fgseaRes_Hallmark_plot %>%
  group_by(pathway) %>%
  filter(!all(is.na(padj))) %>%
  ungroup()

hallmark_bubble_plot <- ggplot(fgseaRes_Hallmark_plot, aes(x = group, y = pathway)) + 
  geom_point(aes(fill = NES, size = -log10(padj)), shape = 21) + 
  labs( x= "", y = "", fill = "NES", size = "-log10(padj)")  + 
  scale_fill_gradient2(low="blue", mid="white", high="red", oob = scales::squish) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", axis.text = element_text(face = "bold", size = 12), 
        legend.text = element_text(face = "bold", size = 10), 
        legend.title = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14) 
  )+
  labs(title = "Gene Set Enrichment Analysis (BTLA)")

svg(paste("./Manuscript_Figures/figure3/HallmarkBTLA_Stim.svg", sep=""), 
    width = 9, height = 10 )
hallmark_bubble_plot
dev.off()


####################
## Motif analysis ##
####################
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

#PD1
DEGs_signif <- DEGs_PD1
for (i in 1:length(DEGs_signif)){DEGs_signif[[i]] <- DEGs_PD1[[i]] %>% dplyr::filter(padj < 0.05)}

# Upregulated
DEGs_signif_up <- DEGs_signif
for (i in 1:length(DEGs_signif_up)){
  DEGs_signif_up[[i]] <- DEGs_signif_up[[i]] %>% 
    dplyr::filter(log2FoldChange > log2(1.5)) %>% rownames(.)
}

motif_PD1_up <- as.list(rep(NA, length(DEGs_signif_up)))
for(i in 1:length(motif_PD1_up)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_up[[i]], colnames(motifRankings))
      motif_PD1_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                      motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_PD1_up[[i]] <- NULL
    })
}
names(motif_PD1_up) <- c("PD1 4h", "PD1 20h", "PD1 48h")
for (i in 1:length(motif_PD1_up)){
  if(length(motif_PD1_up[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_PD1_up[[i]] <- motif_PD1_up[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_PD1_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_PD1_up[[i]]$TF_highConf)
    motif_PD1_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_PD1_up[[i]]$TF_highConf)
    motif_PD1_up[[i]]$TF_highConf <- sub(" .*", "", motif_PD1_up[[i]]$TF_highConf)
    motif_PD1_up[[i]]$TF_highConf <- gsub(";", "", motif_PD1_up[[i]]$TF_highConf)
    motif_PD1_up[[i]]$group <- names(motif_PD1_up)[i]
  }
}


motif_PD1_up <- motif_PD1_up[lapply(motif_PD1_up,length)>1] ## you can use sapply,rapply

write.xlsx(motif_PD1_up, "Manuscript_Figures/others/motif_PD1.xlsx")

motif_PD1_up_combined <- do.call(rbind, motif_PD1_up)
motif_PD1_up_combined <- motif_PD1_up_combined %>% unique(.)

#selection
to_select_PD1_up <- motif_PD1_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_PD1_up_combined <- motif_PD1_up_combined %>% dplyr::filter(TF_highConf %in% to_select_PD1_up)
motif_PD1_up_combined <- motif_PD1_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_PD1_up <- (max(motif_PD1_up_combined$NES) + min(motif_PD1_up_combined$NES))/ 2

# Ordering plot
motif_PD1_up_combined$group <- factor(motif_PD1_up_combined$group, 
                                       levels = c("PD1 4h", "PD1 20h", "PD1 48h"))
motif_PD1_up_combined <- motif_PD1_up_combined[order(motif_PD1_up_combined$group,
                                                       motif_PD1_up_combined$NES),]
TF_order_PD1_up <- motif_PD1_up_combined$TF_highConf %>% unique(.)
motif_PD1_up_combined$TF_highConf <- factor(motif_PD1_up_combined$TF_highConf, 
                                             levels = TF_order_PD1_up)
#plot
motif_PD1_up_plot <- ggplot(motif_PD1_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_PD1_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for upregulated stimulated PD1 genes")
svg(paste("./Manuscript_Figures/figure3/MotifUpregulatedPD1_Stim.svg", sep=""), 
    width = 24, height = 3.5 )
motif_PD1_up_plot
dev.off()

# Downregulated
DEGs_signif_down <- DEGs_signif
for (i in 1:length(DEGs_signif_down)){
  DEGs_signif_down[[i]] <- DEGs_signif_down[[i]] %>% 
    dplyr::filter(log2FoldChange < -log2(1.5)) %>% rownames(.)
}

motif_PD1_down <- as.list(rep(NA, length(DEGs_signif_down)))
for(i in 1:length(motif_PD1_down)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_down[[i]], colnames(motifRankings))
      motif_PD1_down[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                        motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_PD1_down[[i]] <- NULL
    })
}
names(motif_PD1_down) <- c("PD1 4h", "PD1 20h", "PD1 48h")
for (i in 1:length(motif_PD1_down)){
  if(length(motif_PD1_down[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_PD1_down[[i]] <- motif_PD1_down[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_PD1_down[[i]]$TF_highConf <-  gsub('\\.', "", motif_PD1_down[[i]]$TF_highConf)
    motif_PD1_down[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_PD1_down[[i]]$TF_highConf)
    motif_PD1_down[[i]]$TF_highConf <- sub(" .*", "", motif_PD1_down[[i]]$TF_highConf)
    motif_PD1_down[[i]]$TF_highConf <- gsub(";", "", motif_PD1_down[[i]]$TF_highConf)
    motif_PD1_down[[i]]$group <- names(motif_PD1_down)[i]
  }
  
}

motif_PD1_down <- motif_PD1_down[lapply(motif_PD1_down,length)>1] ## you can use sapply,rapply

write.xlsx(motif_PD1_down, "Manuscript_Figures/others/motif_PD1_down.xlsx")

motif_PD1_down_combined <- do.call(rbind, motif_PD1_down)
motif_PD1_down_combined <- motif_PD1_down_combined %>% unique(.)

#selection
to_select_PD1_down <- motif_PD1_down_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_PD1_down_combined <- motif_PD1_down_combined %>% dplyr::filter(TF_highConf %in% to_select_PD1_down)
motif_PD1_down_combined <- motif_PD1_down_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_PD1_down <- (max(motif_PD1_down_combined$NES) + min(motif_PD1_down_combined$NES))/ 2

# Ordering plot
motif_PD1_down_combined$group <- factor(motif_PD1_down_combined$group, 
                                         levels = c("PD1 4h", "PD1 20h", "PD1 48h"))
motif_PD1_down_combined <- motif_PD1_down_combined[order(motif_PD1_down_combined$group,
                                                           motif_PD1_down_combined$NES),]
TF_order_PD1_down <- motif_PD1_down_combined$TF_highConf %>% unique(.)
motif_PD1_down_combined$TF_highConf <- factor(motif_PD1_down_combined$TF_highConf, 
                                               levels = TF_order_PD1_down)
#plot
motif_PD1_down_plot <- ggplot(motif_PD1_down_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_PD1_down) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for downregulated stimulated PD1 genes")
svg(paste("./Manuscript_Figures/figure3/MotifDownregulatedPD1_Stim.svg", sep=""), 
    width = 24, height = 3.5 )
motif_PD1_down_plot
dev.off()

#BTLA
DEGs_signif <- DEGs_BTLA
for (i in 1:length(DEGs_signif)){DEGs_signif[[i]] <- DEGs_BTLA[[i]] %>% dplyr::filter(padj < 0.05)}

# Upregulated
DEGs_signif_up <- DEGs_signif
for (i in 1:length(DEGs_signif_up)){
  DEGs_signif_up[[i]] <- DEGs_signif_up[[i]] %>% 
    dplyr::filter(log2FoldChange > log2(1.5)) %>% rownames(.)
}

motif_BTLA_up <- as.list(rep(NA, length(DEGs_signif_up)))
for(i in 1:length(motif_BTLA_up)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_up[[i]], colnames(motifRankings))
      motif_BTLA_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                      motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_BTLA_up[[i]] <- NULL
    })
}
names(motif_BTLA_up) <- c("BTLA 4h", "BTLA 20h", "BTLA 48h")
for (i in 1:length(motif_BTLA_up)){
  if(length(motif_BTLA_up[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_BTLA_up[[i]] <- motif_BTLA_up[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_BTLA_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_BTLA_up[[i]]$TF_highConf)
    motif_BTLA_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_BTLA_up[[i]]$TF_highConf)
    motif_BTLA_up[[i]]$TF_highConf <- sub(" .*", "", motif_BTLA_up[[i]]$TF_highConf)
    motif_BTLA_up[[i]]$TF_highConf <- gsub(";", "", motif_BTLA_up[[i]]$TF_highConf)
    motif_BTLA_up[[i]]$group <- names(motif_BTLA_up)[i]
  }
}

motif_BTLA_up <- motif_BTLA_up[lapply(motif_BTLA_up,length)>1] ## you can use sapply,rapply

write.xlsx(motif_BTLA_up, "Manuscript_Figures/others/motif_BTLA_up.xlsx")

motif_BTLA_up_combined <- do.call(rbind, motif_BTLA_up)
motif_BTLA_up_combined <- motif_BTLA_up_combined %>% unique(.)

#selection
to_select_BTLA_up <- motif_BTLA_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_BTLA_up_combined <- motif_BTLA_up_combined %>% dplyr::filter(TF_highConf %in% to_select_BTLA_up)
motif_BTLA_up_combined <- motif_BTLA_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_BTLA_up <- (max(motif_BTLA_up_combined$NES) + min(motif_BTLA_up_combined$NES))/ 2

# Ordering plot
motif_BTLA_up_combined$group <- factor(motif_BTLA_up_combined$group, 
                                       levels = c("BTLA 4h", "BTLA 20h", "BTLA 48h"))
motif_BTLA_up_combined <- motif_BTLA_up_combined[order(motif_BTLA_up_combined$group,
                                                       motif_BTLA_up_combined$NES),]
TF_order_BTLA_up <- motif_BTLA_up_combined$TF_highConf %>% unique(.)
motif_BTLA_up_combined$TF_highConf <- factor(motif_BTLA_up_combined$TF_highConf, 
                                             levels = TF_order_BTLA_up)
#plot
motif_BTLA_up_plot <- ggplot(motif_BTLA_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_BTLA_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for upregulated stimulated BTLA genes")
svg(paste("./Manuscript_Figures/figure3/MotifUpregulatedBTLA_Stim.svg", sep=""), 
    width = 24, height = 3.5 )
motif_BTLA_up_plot
dev.off()

# Downregulated
DEGs_signif_down <- DEGs_signif
for (i in 1:length(DEGs_signif_down)){
  DEGs_signif_down[[i]] <- DEGs_signif_down[[i]] %>% 
    dplyr::filter(log2FoldChange < -log2(1.5)) %>% rownames(.)
}

motif_BTLA_down <- as.list(rep(NA, length(DEGs_signif_down)))
for(i in 1:length(motif_BTLA_down)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_down[[i]], colnames(motifRankings))
      motif_BTLA_down[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                        motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_BTLA_down[[i]] <- NULL
    })
}
names(motif_BTLA_down) <- c("BTLA 4h", "BTLA 20h", "BTLA 48h")
for (i in 1:length(motif_BTLA_down)){
  if(length(motif_BTLA_down[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_BTLA_down[[i]] <- motif_BTLA_down[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_BTLA_down[[i]]$TF_highConf <-  gsub('\\.', "", motif_BTLA_down[[i]]$TF_highConf)
    motif_BTLA_down[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_BTLA_down[[i]]$TF_highConf)
    motif_BTLA_down[[i]]$TF_highConf <- sub(" .*", "", motif_BTLA_down[[i]]$TF_highConf)
    motif_BTLA_down[[i]]$TF_highConf <- gsub(";", "", motif_BTLA_down[[i]]$TF_highConf)
    motif_BTLA_down[[i]]$group <- names(motif_BTLA_down)[i]
  }
  
}

motif_BTLA_down <- motif_BTLA_down[lapply(motif_BTLA_down,length)>1] ## you can use sapply,rapply

write.xlsx(motif_BTLA_down, "Manuscript_Figures/others/motif_BTLA_down.xlsx")

motif_BTLA_down_combined <- do.call(rbind, motif_BTLA_down)
motif_BTLA_down_combined <- motif_BTLA_down_combined %>% unique(.)

#selection
to_select_BTLA_down <- motif_BTLA_down_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_BTLA_down_combined <- motif_BTLA_down_combined %>% dplyr::filter(TF_highConf %in% to_select_BTLA_down)
motif_BTLA_down_combined <- motif_BTLA_down_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_BTLA_down <- (max(motif_BTLA_down_combined$NES) + min(motif_BTLA_down_combined$NES))/ 2

# Ordering plot
motif_BTLA_down_combined$group <- factor(motif_BTLA_down_combined$group, 
                                         levels = c("BTLA 4h", "BTLA 20h", "BTLA 48h"))
motif_BTLA_down_combined <- motif_BTLA_down_combined[order(motif_BTLA_down_combined$group,
                                                           motif_BTLA_down_combined$NES),]
TF_order_BTLA_down <- motif_BTLA_down_combined$TF_highConf %>% unique(.)
motif_BTLA_down_combined$TF_highConf <- factor(motif_BTLA_down_combined$TF_highConf, 
                                               levels = TF_order_BTLA_down)
#plot
motif_BTLA_down_plot <- ggplot(motif_BTLA_down_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_BTLA_down) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for downregulated stimulated BTLA genes")
svg(paste("./Manuscript_Figures/figure3/MotifDownregulatedBTLA_Stim.svg", sep=""), 
    width = 24, height = 3.5 )
motif_BTLA_down_plot
dev.off()



################################
## Kinase Enrichment Analysis ##
################################
library(httr)
library(jsonlite)


#BTLA
DEGs.selection.down <- lapply(DEGs_BTLA, subset, padj < 0.05 & log2FoldChange < -log2(1.5))

for (i in 1:length(DEGs.selection.down)){
  DEGs.selection.down[[i]] <- DEGs.selection.down[[i]] %>% 
  arrange(padj) %>%
  head(., 300)
}
DEGs.selection.down <- lapply(DEGs.selection.down, rownames)

KEA_results <- as.list(rep(NA, length(DEGs_BTLA)))
names(KEA_results) <- names(DEGs_BTLA)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.down[[i]]

  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_BTLA.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_BTLA.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("4h",1), rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("4h", "20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("4h", "20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
# col <- colorRampPalette(brewer.pal(10, "YlOrRd"))(256)
# col <- rev(col)
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                              col = col_fun,top_annotation = ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "black",
                              column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

svg(paste0("./Manuscript_Figures/others/KEA_BTLA_Heatmap.svg"), 
    width = 7, height = 4.5 )
print(ComplexHeatmap::Heatmap(matrix = (new_df), 
                              col = col_fun,top_annotation = ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("KEA3 Rank Score - BTLA - MeanRank"), 
                              name = " "))
dev.off()




#PD1
DEGs.selection.down <- lapply(DEGs_PD1, subset, padj < 0.05 & log2FoldChange < -log2(1.5))

for (i in 1:length(DEGs.selection.down)){
  DEGs.selection.down[[i]] <- DEGs.selection.down[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.down <- lapply(DEGs.selection.down, rownames)

KEA_results <- as.list(rep(NA, length(DEGs_PD1)))
names(KEA_results) <- names(DEGs_PD1)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.down[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_PD1.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_PD1.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("4h",1), rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("4h", "20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("4h", "20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
# col <- colorRampPalette(brewer.pal(10, "YlOrRd"))(256)
# col <- rev(col)
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

svg(paste0("./Manuscript_Figures/others/KEA_PD1_Heatmap.svg"), 
    width = 7, height = 4.5 )
print(ComplexHeatmap::Heatmap(matrix = (new_df), 
                              col = col_fun,top_annotation = ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("KEA3 Rank Score - PD1 - MeanRank"), 
                              name = " "))
dev.off()




################################
## Kinase Enrichment Analysis ##
################################
#PD1
DEGs.selection.up <- lapply(DEGs_PD1, subset, padj < 0.05 & log2FoldChange < -log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_PD1.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_PD1.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("4h",1), rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("4h", "20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("4h", "20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

ha1 = HeatmapAnnotation(type = type$group,name = " ",
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")), which = "row", annotation_label = " ")

svg(paste0("./Manuscript_Figures/figure3/KEA_PD1_Heatmap.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun, left_annotation =  ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()



#BTLA
DEGs.selection.up <- lapply(DEGs_BTLA, subset, padj < 0.05 & log2FoldChange < -log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_BTLA.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_BTLA.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("4h","20h","48h")

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("4h",1), rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("4h", "20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("4h", "20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

ha1 = HeatmapAnnotation(type = type$group,name = " ",
                        col = list(type = c("4h" = "#198482", "20h" = "#F49C40", 
                                            "48h" = "red")), which = "row", annotation_label = " ")

svg(paste0("./Manuscript_Figures/figure3/KEA_BTLA_Heatmap.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun, left_annotation =  ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()
```




# Correlation analysis

```{r}
DEGs <- list()
DEGs[["StimvsUnstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "ctrl_4h", "unstimuliert_0"), 
                                                  format = "DataFrame"))
DEGs[["StimvsUnstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_20h", "unstimuliert_0"), 
                                                   format = "DataFrame"))
DEGs[["StimvsUnstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "ctrl_48h", "unstimuliert_0"), 
                                                   format = "DataFrame"))


DEGs_PD1 <- list()

DEGs_PD1[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))

DEGs_BTLA <- list()

DEGs_BTLA[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                          contrast = c("Remarks", "BTLA_4h", "ctrl_4h"), 
                                                          format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                           format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                           format = "DataFrame"))


DEGs_TCR_cor_up <- DEGs
DEGs_TCR_cor_down <- DEGs
DEGs_PD1_cor_up <- DEGs_PD1
DEGs_PD1_cor_down <- DEGs_PD1
DEGs_BTLA_cor_up <- DEGs_BTLA
DEGs_BTLA_cor_down <- DEGs_BTLA

# Upregulated signif TCR
result_upregulated <- as.list(rep(NA, length(DEGs)))
names(result_upregulated) <- c("4h", "20h", "48h")

for (i in 1:length(DEGs)){
  temp <- DEGs[[i]] %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5) ) %>%
    rownames(.)
  temp2 <- DEGs_PD1[[i]] %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5) ) %>%
    rownames(.)
  temp3 <- DEGs_BTLA[[i]] %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5) ) %>%
    rownames(.)
  row_name_inhib <- unique(c(temp2,temp3))
  row_name_all <- intersect(row_name_inhib, temp)

  DEGs_TCR_cor_up[[i]] <- DEGs_TCR_cor_up[[i]][row_name_all,]%>%
    dplyr::select(stat) %>%
    dplyr::rename(TCR_stat = stat)
  DEGs_PD1_cor_up[[i]] <- DEGs_PD1_cor_up[[i]][row_name_all,]%>%
    dplyr::select(stat) %>%
    dplyr::rename(PD1_stat = stat)
  DEGs_BTLA_cor_up[[i]] <- DEGs_BTLA_cor_up[[i]][row_name_all,]%>%
    dplyr::select(stat) %>%
    dplyr::rename(BTLA_stat = stat)
  
  result_upregulated[[i]] <- do.call("cbind", list(DEGs_TCR_cor_up[[i]], 
                                                DEGs_PD1_cor_up[[i]],
                                                DEGs_BTLA_cor_up[[i]]))
  
  result_upregulated[[i]]$gene <- rownames(result_upregulated[[i]])
}


FACS_list <- c(NULL)
for (i in 1:length(result_upregulated)){
  
  annot_tbl_PD1 <- result_upregulated[[i]] %>% 
    arrange(PD1_stat) %>%
    ht(m = 10, n = 10) %>%
    pull(gene)
  
  plot_PD1 <- ggplot(data = result_upregulated[[i]], mapping = aes(x = TCR_stat, y= PD1_stat)) +
    geom_point(mapping = aes(color = ifelse(result_upregulated[[i]]$PD1_stat > 0, 
                                            "PD1 Upregulated","PD1 Downregulated"))) +
    ggrepel::geom_text_repel(aes(label=ifelse(gene %in% annot_tbl_PD1 | gene %in% FACS_list,
                                              as.character(gene),'') ,
                                 color = ifelse(gene %in% FACS_list, "Validated with FACS",
                                                ifelse(result_upregulated[[i]]$PD1_stat > 0,
                                                       "PD1 Upregulated","PD1 Downregulated"))),
                             show.legend=FALSE,max.overlaps = 100, 
                             hjust=0,vjust=0) +
    theme_ipsum() +
    theme(axis.title.x  = element_text(size = 16, family = "arial"),
          axis.title.y = element_text(size = 16, family = "arial"),
          legend.text = element_text(size = 18, family = "arial")) +
    labs(title = paste0("PD-1 ",names(result_upregulated[i])), 
         x = "TCR stat score", 
         y= "PD-1 stat score",
         color=" ")
  
  
  annot_tbl_BTLA <- result_upregulated[[i]] %>% 
    arrange(BTLA_stat) %>%
    ht(m = 10, n = 10) %>%
    pull(gene)
  
  plot_BTLA <- ggplot(data = result_upregulated[[i]], mapping = aes(x = TCR_stat, y= BTLA_stat)) +
    geom_point(mapping = aes(color = ifelse(result_upregulated[[i]]$BTLA_stat > 0, 
                                          "BTLA Upregulated","BTLA Downregulated"))) +
    ggrepel::geom_text_repel(aes(label=ifelse(gene %in% annot_tbl_BTLA | gene %in% FACS_list,
                                              as.character(gene),'') ,
                                 color = ifelse(gene %in% FACS_list, "Validated with FACS",
                                                ifelse(result_upregulated[[i]]$BTLA_stat > 0,
                                                       "BTLA Upregulated","BTLA Downregulated"))),
                             show.legend=FALSE,max.overlaps = 100, 
                             hjust=0,vjust=0) +
    theme_ipsum() +
    theme(axis.title.x  = element_text(size = 16, family = "arial"),
          axis.title.y = element_text(size = 16, family = "arial"), 
          legend.text = element_text(size = 18, family = "arial")) +
    labs(title = paste0("BTLA ",names(result_upregulated[i])), 
       x = "TCR stat score", 
       y= "BTLA stat score",
       color=" ")
  
  
  svg(paste0("./Manuscript_Figures/figure3/Cor_BTLA_",names(result_upregulated)[i],".svg"), 
    width = 12, height = 8 )
  print(plot_BTLA)
  dev.off()
  
  svg(paste0("./Manuscript_Figures/figure3/Cor_PD1_",names(result_upregulated)[i],".svg"), 
  width = 12, height = 8 )
  print(plot_PD1)
  dev.off()
}
```

## Figure 4: scRNA-seq TCR

```{r}
###################
### Upregulated ###
###################

#### MFuzz #### 
exp_design_TCR <- exp_design %>% filter(str_detect(Remarks, c("unstimuliert|ctrl_"))) %>% rownames(.)
DEGs_TCR <- DEGs
for(i in 1:length(DEGs)){
  DEGs_TCR[[i]] <- DEGs_TCR[[i]] %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > log2(1.5)) %>% 
    rownames()
}
DEGs_TCR <- Reduce(c,DEGs_TCR) %>% unique(.)
normalized_counts_TCR_up <- normalized_counts[DEGs_TCR, exp_design_TCR]

# Aggregation and MFuzz clustering
normalized_counts_TCR_up <- as.data.frame(t(normalized_counts_TCR_up))
normalized_counts_TCR_up$stim <- c(rep("unstimulated",3), rep("stimulated 4h",3), 
                                   rep("stimulated 20h",3), rep("stimulated 48h", 3) )
normalized_counts_TCR_up <- normalized_counts_TCR_up%>%
  group_by(stim)%>% 
  summarise(across(everything(), list(mean)))
normalized_counts_TCR_up <- normalized_counts_TCR_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="stim")
normalized_counts_TCR_up <- as.data.frame(t(normalized_counts_TCR_up))
normalized_counts_TCR_up$gene <- rownames(normalized_counts_TCR_up)
normalized_counts_TCR_up$gene <- gsub('.{2}$', '', normalized_counts_TCR_up$gene)
normalized_counts_TCR_up <- normalized_counts_TCR_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="gene")
col_order <- c("unstimulated", "stimulated 4h","stimulated 20h", "stimulated 48h")
normalized_counts_TCR_up <- normalized_counts_TCR_up[, col_order]

eset_TCR <- new("ExpressionSet",exprs=as.matrix(normalized_counts_TCR_up))

eset_TCR <- Mfuzz::filter.NA(eset_TCR, thres=0.25)
eset_TCR <- Mfuzz::fill.NA(eset_TCR,mode="mean")
eset_TCR <- Mfuzz::standardise(eset_TCR)
m1 <- mestimate(eset_TCR)
n_clusters <- 9

set.seed(1)
cl_4 <- mfuzz(eset_TCR,c=n_clusters,m=m1)
svg(paste("./Manuscript_Figures/figure4/UP_Mfuzz_TCR.svg", sep=""), 
    width = 14, height = 14 )
mfuzz.plot2(eset_TCR,cl=cl_4,mfrow=c(3,3), x11 = FALSE , time.labels = col_order )
dev.off()
CLUSTERlistTCR <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistTCR <- split(CLUSTERlistTCR, CLUSTERlistTCR$cluster)



#KEA3
for(i in 1:length(CLUSTERlistTCR)){
  CLUSTERlistTCR[[i]] <- CLUSTERlistTCR[[i]]$gene
}

KEA_results <- as.list(rep(NA, length(CLUSTERlistTCR)))
names(KEA_results) <- names(CLUSTERlistTCR)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = CLUSTERlistTCR[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_TCR_MFuzz.xlsx")


# KEA visualization
top_10 <- lapply(KEA_results, function(i) head(i, 10)) 
for (i in 1:length(top_10)){top_10[[i]] <- top_10[[i]] %>% pull(TF)}
top_10 <- Reduce(c,top_10) %>% unique(.)
KEA_results_subset <- KEA_results


for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% dplyr::filter(TF %in% top_10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- paste0("cluster",i)
}

to_plot <- bind_rows(KEA_results_subset)
#to_plot$group <- factor(to_plot$group, levels = c("4h", "20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)

pathway_order <- to_plot %>% dplyr::filter(group %in% "cluster1") %>%
  arrange(desc(logScore)) %>%
  pull(TF)
to_plot$TF <- factor(to_plot$TF, levels = pathway_order )
to_plot$logScore <- ifelse(to_plot$Score < 100, to_plot$logScore, NA)


KEA_bubble_plot <- ggplot(to_plot, aes(x = group, y = TF)) + 
  geom_point(aes(color = logScore ), size = 5) + 
  labs( x= "", y = "",size = "Log-score")  + 
  scale_color_gradient2(low="yellow", mid = "orange",high="red", midpoint = mean(to_plot$logScore, na.rm = T),na.value = "#FFFFFF" ) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", 
        axis.text = element_text(face = "bold", size = 16), 
        legend.text = element_text(face = "bold", size = 16), 
        legend.title = element_text(face = "bold", size = 16),
        title = element_text(face = "bold", size = 22) 
  )+
  labs(title = "Kinase Enrichment Analysis (TCR)")

svg(paste("./Manuscript_Figures/figure4/KEA_TCR_MFuzz_Plot.svg", sep=""), 
    width = 15, height = 15 )
KEA_bubble_plot
dev.off()


#### KEGG and REACTOME ORA enrichment analysis ####
CLUSTERlistTCR <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistTCR <- split(CLUSTERlistTCR, CLUSTERlistTCR$cluster)

for(i in 1:length(CLUSTERlistTCR)){CLUSTERlistTCR[[i]] <- CLUSTERlistTCR[[i]] %>% dplyr::select(-cluster)}

pwsTCR <- gost(CLUSTERlistTCR, organism = "hsapiens", correction_method = "gSCS",
             user_threshold = 0.05, evcodes = TRUE, ordered_query = FALSE,
             sources = c("KEGG","REAC"))$result

pwsTCR <- pwsTCR[order(as.numeric(pwsTCR$query)),]
pwsTCR$p_value <- -log10(pwsTCR$p_value) #transform to -log10(pval)
pwsTCR.r <- reshape(pwsTCR[,c(1,3,9,10,11)], timevar = "query", idvar = c("term_id","source","term_name"), direction = "wide")
colnames(pwsTCR.r) <- gsub("p_value.","",colnames(pwsTCR.r))
write.csv(pwsTCR.r, "./Manuscript_Figures/figure4/UP_GO_ORA_Mfuzz_TCR.csv")
saveRDS(CLUSTERlistTCR, "./Manuscript_Figures/figure4/UP_Mfuzz_genes_TCR.rds")

pwsTCR_plot <- pwsTCR
pwsTCR_plot$query <- paste0("cluster",pwsTCR_plot$query)

#selection
to_select <- pwsTCR_plot %>% 
  arrange(desc(p_value)) %>% 
  group_by(query) %>% dplyr::slice(1:15) %>%
  pull(term_name) %>%
  unique(.)
pwsTCR_plot <- pwsTCR_plot %>% dplyr::filter(term_name %in% to_select)
pwsTCR_plot <- pwsTCR_plot %>% distinct(term_name, query, .keep_all = TRUE)
midpoint <- (max(pwsTCR_plot$p_value) + min(pwsTCR_plot$p_value))/ 2

# # Ordering plot
pwsTCR_plot$query <- as.factor(pwsTCR_plot$query)
pwsTCR_plot <- pwsTCR_plot[order(pwsTCR_plot$query,
                                 pwsTCR_plot$p_value),]
term_order <- pwsTCR_plot$term_name %>% unique(.)
pwsTCR_plot$term_name <- factor(pwsTCR_plot$term_name,
                                             levels = term_order)
#plot
svg(paste("./Manuscript_Figures/figure4/PathwayUpregulated_TCR_MFuzz.svg", sep=""), 
    width = 24, height = 20 )
ggplot(pwsTCR_plot, 
       aes(x = query, y = term_name)) + 
  geom_point(aes(fill = p_value), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "p_value")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(vjust = 0.5, hjust=0.5, size = 24, face = "bold"),
        axis.text.y = element_text(size = 24, face = "bold"),
        plot.title  = element_text(size = 30, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "Gene sets enrichment analysis for MFuzz clusters of\n genes upregulated upon TCR-complex stimulation")
dev.off()

#### Motif analysis ####
CLUSTERlistTCR <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistTCR <- split(CLUSTERlistTCR, CLUSTERlistTCR$cluster)

for(i in 1:length(CLUSTERlistTCR)){
  CLUSTERlistTCR[[i]] <- CLUSTERlistTCR[[i]] %>% 
  dplyr::select(-cluster) %>% pull(.)
} 
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

motif_TCR_MFuzz_up <- as.list(rep(NA, length(CLUSTERlistTCR)))
for(i in 1:length(motif_TCR_MFuzz_up)){
  tryCatch(
    expr = {
      temp <- intersect(CLUSTERlistTCR[[i]], colnames(motifRankings))
      motif_TCR_MFuzz_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                      motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_TCR_MFuzz_up[[i]] <- NULL
    })
}
names(motif_TCR_MFuzz_up) <- paste0("cluster",1:length(motif_TCR_MFuzz_up))
for (i in 1:length(motif_TCR_MFuzz_up)){
  if(length(motif_TCR_MFuzz_up[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_TCR_MFuzz_up[[i]] <- motif_TCR_MFuzz_up[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_TCR_MFuzz_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_TCR_MFuzz_up[[i]]$TF_highConf)
    motif_TCR_MFuzz_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_TCR_MFuzz_up[[i]]$TF_highConf)
    motif_TCR_MFuzz_up[[i]]$TF_highConf <- sub(" .*", "", motif_TCR_MFuzz_up[[i]]$TF_highConf)
    motif_TCR_MFuzz_up[[i]]$TF_highConf <- gsub(";", "", motif_TCR_MFuzz_up[[i]]$TF_highConf)
    motif_TCR_MFuzz_up[[i]]$group <- names(motif_TCR_MFuzz_up)[i]
  }
}

motif_TCR_MFuzz_up <- motif_TCR_MFuzz_up[lapply(motif_TCR_MFuzz_up,length)>1] ## you can use sapply,rapply
motif_TCR_MFuzz_up_combined <- do.call(rbind, motif_TCR_MFuzz_up)
motif_TCR_MFuzz_up_combined <- motif_TCR_MFuzz_up_combined %>% unique(.)

#selection
to_select_TCR_MFuzz_up <- motif_TCR_MFuzz_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:15) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_TCR_MFuzz_up_combined <- motif_TCR_MFuzz_up_combined %>% dplyr::filter(TF_highConf %in% to_select_TCR_MFuzz_up)
motif_TCR_MFuzz_up_combined <- motif_TCR_MFuzz_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_PD1_up <- (max(motif_TCR_MFuzz_up_combined$NES) + min(motif_TCR_MFuzz_up_combined$NES))/ 2

# # Ordering plot
motif_TCR_MFuzz_up_combined$group <- as.factor(motif_TCR_MFuzz_up_combined$group)
motif_TCR_MFuzz_up_combined <- motif_TCR_MFuzz_up_combined[order(motif_TCR_MFuzz_up_combined$group,
                                                       motif_TCR_MFuzz_up_combined$NES),]
TF_order <- motif_TCR_MFuzz_up_combined$TF_highConf %>% unique(.)
motif_TCR_MFuzz_up_combined$TF_highConf <- factor(motif_TCR_MFuzz_up_combined$TF_highConf,
                                             levels = TF_order)
#plot
motif_TCR_up_plot <- ggplot(motif_TCR_MFuzz_up_combined, aes(x = group, y = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_PD1_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        plot.title = element_text(size = 34, face = "bold"),
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for MFuzz clusters\nof genes upregulated upon TCR-complex stimulation")

svg(paste("./Manuscript_Figures/figure4/MotifUpregulated_TCR_MFuzz.svg", sep=""), 
    width = 12, height = 28 )
motif_TCR_up_plot
dev.off()


################################
## Kinase Enrichment Analysis ##
################################
#TCR
DEGs.selection.up <- CLUSTERlistTCR

KEA_results <- as.list(rep(NA, length(DEGs.selection.up)))
names(KEA_results) <- names(DEGs.selection.up)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_TCR_MFuzz.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_TCR_MFuzz.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- 1:9

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = 1:9 )
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = 1:9)
type$group <- factor(type$group, levels = 1:9)
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("1" = "#198482", "2" = "#F49C40", 
                                            "3" = "red"," 4" = "blue","5" = "green",
                                            "6" = "purple", "7" = "pink", "8" = "grey", "9" = "orange")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- 1:9
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA
colnames(new_df) <- paste("Cluster", colnames(new_df), sep = "_")

svg(paste0("./Manuscript_Figures/figure4/KEA_TCR_Heatmap_Mfuzz.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun,
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()


# activated T-cell data
activated_seurat <- readRDS("../activated_T_cells_seurat.rds")
activated_seurat_metadata <- activated_seurat@meta.data
activated_seurat_metadata_patient <- activated_seurat_metadata %>% dplyr::select(Patient) %>% unique() %>% data.frame()
activated_seurat_metadata_patient$donor <- c("donor1","donor1","donor1","donor1","donor1","donor1",
                                             "donor2","donor2","donor2","donor2","donor2","donor2",
                                             "donorA","donorA","donorB","donorB")
activated_seurat_metadata <- left_join(activated_seurat_metadata, activated_seurat_metadata_patient)
activated_seurat$Donor <- activated_seurat_metadata$donor
activated_seurat_donor <- Seurat::SplitObject(activated_seurat, split.by = "Tissue")

activated_seurat_list <- readRDS("../activated_T_cells_seurat_list.rds")

for (i in 1:length(activated_seurat_list)){
  activated_seurat_list[[i]]$Donor <- activated_seurat_donor[[i]]$Donor
}

activated_seurat_Blood <- activated_seurat_list$Blood

# cell type annotation
lapply(c("dplyr","Seurat","HGNChelper"), library, character.only = T)
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"); source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")

# get cell-type-specific gene sets from our in-built database (DB)
gs_list = gene_sets_prepare("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx", "Immune system")
gs_list$gs_positive <-  gs_list$gs_positive[grep("T cells", names(gs_list$gs_positive))]
gs_list$gs_negative <-  gs_list$gs_negative[grep("T cells", names(gs_list$gs_negative))]

# assign cell types
es.max <- sctype_score(scRNAseqData = as.matrix(activated_seurat_Blood@assays$RNA@counts), 
                       scaled = F, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
es.max <- as.data.frame(es.max)
es.max.type <- rownames(es.max)[apply(es.max,2,which.max)]

activated_seurat_Blood$celltype_individual <- es.max.type
activated_seurat_Blood$celltype <- es.max.type
# merge by cluster
cL_resutls = do.call("rbind", lapply(unique(activated_seurat_Blood@meta.data$seurat_clusters), function(cl){
    es.max.cl = sort(rowSums(es.max[ ,rownames(activated_seurat_Blood@meta.data[activated_seurat_Blood@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
    head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(activated_seurat_Blood@meta.data$seurat_clusters==cl)), 10)
}))
sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  

activated_seurat_Blood@meta.data$celltype = ""
for(j in unique(sctype_scores$cluster)){
  cl_type = sctype_scores[sctype_scores$cluster==j,]; 
  activated_seurat_Blood@meta.data$celltype[activated_seurat_Blood@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])
}


svg(paste("./Manuscript_Figures/figure4/UMAP_scRNAseq_state.svg", sep=""), 
    width = 8, height = 6 )
DimPlot(activated_seurat_Blood, group.by = "State") +
  theme(axis.title = element_text(size = 27), 
        legend.text = element_text(size = 27), 
        title = element_text(size = 27))+ 
  guides(colour = guide_legend(override.aes = list(size=10)))
dev.off()
svg(paste("./Manuscript_Figures/figure4/UMAP_scRNAseq_celltype.svg", sep=""), 
    width = 10, height = 6 )
DimPlot(activated_seurat_Blood, group.by = "celltype") + labs(title = "Cell type") +
  theme(axis.title = element_text(size = 27), 
        legend.text = element_text(size = 27), 
        title = element_text(size = 27))+ 
  guides(colour = guide_legend(override.aes = list(size=10)))
dev.off()
svg(paste("./Manuscript_Figures/figure4/UMAP_scRNAseq_Donor.svg", sep=""), 
    width = 8, height = 6 )
DimPlot(activated_seurat_Blood, group.by = "Donor") +
  theme(axis.title = element_text(size = 27), 
        legend.text = element_text(size = 27), 
        title = element_text(size = 27))+ 
  guides(colour = guide_legend(override.aes = list(size=10)))
dev.off()


#### Stimulated T-cells sc scoring ####
library(UCell)
CLUSTERlistTCR <- readRDS("./Manuscript_Figures/figure4/UP_Mfuzz_genes_TCR.rds")

for (i in 1:length(CLUSTERlistTCR)){
  cluster_MFuzz_TCR <-  rbind(CLUSTERlistTCR[[i]] )%>% 
    pull(.)
  
  activated_seurat_Blood <- AddModuleScore_UCell(
    obj  = activated_seurat_Blood,
    features = list(cluster_MFuzz_TCR), name = paste0("cluster",i)
    )
}
plot_TCR_upregulated <- as.list(rep(NA, length(CLUSTERlistTCR)))
for (i in 1:length(CLUSTERlistTCR)){
  plot_TCR_upregulated[[i]] <- FeaturePlot(activated_seurat_Blood, 
                                           features = paste0("signature_1cluster",i) )+
    scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))+
    labs(title = paste("TCR Signature MFuzz Cluster",i))
    
}
library(cowplot)
svg(paste("./Manuscript_Figures/figure4/UP_Mfuzz_UMAP_TCR.svg", sep=""), 
    width = 14, height = 14 )
plot_grid(plot_TCR_upregulated[[1]],plot_TCR_upregulated[[2]],
          plot_TCR_upregulated[[3]],plot_TCR_upregulated[[4]],
          plot_TCR_upregulated[[5]],plot_TCR_upregulated[[6]],
          plot_TCR_upregulated[[7]],plot_TCR_upregulated[[8]],
          plot_TCR_upregulated[[9]],
          labels = NULL)
dev.off()
```


## Figure 5: scRNA-seq PD1 and BTLA

```{r}
##########
## PD1 ##
##########

#####################
### Downregulated ###
#####################

#### MFuzz #### 
exp_design_PD1 <- exp_design %>% filter(str_detect(Remarks, c("PD1_"))) %>% rownames(.)

DEGs_PD1 <- list()

DEGs_PD1[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))

for(i in 1:length(DEGs)){
  DEGs_PD1[[i]] <- DEGs_PD1[[i]] %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -log2(1.5)) %>% 
    rownames()
}
DEGs_PD1 <- Reduce(c,DEGs_PD1) %>% unique(.)
normalized_counts_PD1_up <- normalized_counts[DEGs_PD1, exp_design_PD1]

# Aggregation and MFuzz clustering
normalized_counts_PD1_up <- as.data.frame(t(normalized_counts_PD1_up))
normalized_counts_PD1_up$stim <- c(rep("PD1 4h",3), 
                                   rep("PD1 20h",3), rep("PD1 48h", 3) )
normalized_counts_PD1_up <- normalized_counts_PD1_up%>%
  group_by(stim)%>% 
  summarise(across(everything(), list(mean)))
normalized_counts_PD1_up <- normalized_counts_PD1_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="stim")
normalized_counts_PD1_up <- as.data.frame(t(normalized_counts_PD1_up))
normalized_counts_PD1_up$gene <- rownames(normalized_counts_PD1_up)
normalized_counts_PD1_up$gene <- gsub('.{2}$', '', normalized_counts_PD1_up$gene)
normalized_counts_PD1_up <- normalized_counts_PD1_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="gene")
col_order <- c( "PD1 4h","PD1 20h", "PD1 48h")
normalized_counts_PD1_up <- normalized_counts_PD1_up[, col_order]


# TCR
exp_design_TCR_PD1 <- exp_design %>% filter(str_detect(Remarks, c("ctrl_"))) %>% rownames(.)
normalized_counts_TCR_PD1 <- normalized_counts[rownames(normalized_counts_PD1_up), exp_design_TCR_PD1]

# Aggregation and MFuzz clustering
normalized_counts_TCR_PD1 <- as.data.frame(t(normalized_counts_TCR_PD1))
normalized_counts_TCR_PD1$stim <- c(rep("stimulated 4h",3), 
                                   rep("stimulated 20h",3), rep("stimulated 48h", 3) )
normalized_counts_TCR_PD1 <- normalized_counts_TCR_PD1%>%
  group_by(stim)%>% 
  summarise(across(everything(), list(mean)))
normalized_counts_TCR_PD1 <- normalized_counts_TCR_PD1 %>% 
  remove_rownames %>% 
  column_to_rownames(var="stim")
normalized_counts_TCR_PD1 <- as.data.frame(t(normalized_counts_TCR_PD1))
normalized_counts_TCR_PD1$gene <- rownames(normalized_counts_TCR_PD1)
normalized_counts_TCR_PD1$gene <- gsub('.{2}$', '', normalized_counts_TCR_PD1$gene)
normalized_counts_TCR_PD1 <- normalized_counts_TCR_PD1 %>% 
  remove_rownames %>% 
  column_to_rownames(var="gene")
col_order <- c("stimulated 4h", "stimulated 20h", "stimulated 48h")
normalized_counts_TCR_PD1 <- normalized_counts_TCR_PD1[, col_order]

# Substraction
normalized_counts_PD1_up <- normalized_counts_PD1_up - normalized_counts_TCR_PD1


eset_PD1 <- new("ExpressionSet",exprs=as.matrix(normalized_counts_PD1_up))

eset_PD1 <- Mfuzz::filter.NA(eset_PD1, thres=0.25)
eset_PD1 <- Mfuzz::fill.NA(eset_PD1,mode="mean")
eset_PD1 <- Mfuzz::standardise(eset_PD1)
m1 <- mestimate(eset_PD1)
n_clusters <- 4

set.seed(1)
cl_4 <- mfuzz(eset_PD1,c=n_clusters,m=m1)
# cl_4_PD1 <- mfuzz(eset_PD1,c=n_clusters,m=m1)

col_order <- c( "PD1 4h","PD1 20h", "PD1 48h")
svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_PD1.svg", sep=""), 
    width = 14, height = 14 )
mfuzz.plot2(eset_PD1,cl=cl_4,mfrow=c(2,2), x11 = FALSE , time.labels = col_order )
dev.off()
CLUSTERlistPD1 <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistPD1 <- split(CLUSTERlistPD1, CLUSTERlistPD1$cluster)


#KEA3 PD1
for(i in 1:length(CLUSTERlistPD1)){
  CLUSTERlistPD1[[i]] <- CLUSTERlistPD1[[i]]$gene
}

KEA_results <- as.list(rep(NA, length(CLUSTERlistPD1)))
names(KEA_results) <- names(CLUSTERlistPD1)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = CLUSTERlistPD1[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_PD1_MFuzz.xlsx")


# KEA visualization
top_10 <- lapply(KEA_results, function(i) head(i, 10)) 
for (i in 1:length(top_10)){top_10[[i]] <- top_10[[i]] %>% pull(TF)}
top_10 <- Reduce(c,top_10) %>% unique(.)
KEA_results_subset <- KEA_results


for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% dplyr::filter(TF %in% top_10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- paste0("cluster",i)
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$Score <- as.numeric(to_plot$Score)

pathway_order <- to_plot %>% dplyr::filter(group %in% "cluster1") %>%
  arrange(desc(logScore)) %>%
  pull(TF)
to_plot$TF <- factor(to_plot$TF, levels = pathway_order )
to_plot$logScore <- ifelse(to_plot$Score < 100, to_plot$logScore, NA)


KEA_bubble_plot <- ggplot(to_plot, aes(x = group, y = TF)) + 
  geom_point(aes(color = logScore ), size = 5) + 
  labs( x= "", y = "",size = "Log-score")  + 
  scale_color_gradient2(low="yellow", mid = "orange",high="red", midpoint = mean(to_plot$logScore, na.rm = T),na.value = "#FFFFFF" ) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", 
        axis.text = element_text(face = "bold", size = 16), 
        legend.text = element_text(face = "bold", size = 16), 
        legend.title = element_text(face = "bold", size = 16),
        title = element_text(face = "bold", size = 22) 
  )+
  labs(title = "Kinase Enrichment Analysis (PD1)")

svg(paste("./Manuscript_Figures/figure5/KEA_PD1_MFuzz_Plot.svg", sep=""), 
    width = 10, height = 10 )
KEA_bubble_plot
dev.off()





#### KEGG and REACTOME ORA enrichment analysis ####
CLUSTERlistPD1 <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistPD1 <- split(CLUSTERlistPD1, CLUSTERlistPD1$cluster)

for(i in 1:length(CLUSTERlistPD1)){CLUSTERlistPD1[[i]] <- CLUSTERlistPD1[[i]] %>%
  dplyr::select(-cluster)}

pwsPD1 <- gost(CLUSTERlistPD1, organism = "hsapiens", correction_method = "gSCS",
               user_threshold = 0.05, evcodes = TRUE, ordered_query = FALSE,
               sources = c("KEGG","REAC"))$result

pwsPD1 <- pwsPD1[order(as.numeric(pwsPD1$query)),]
pwsPD1$p_value <- -log10(pwsPD1$p_value) #transform to -log10(pval)
pwsPD1.r <- reshape(pwsPD1[,c(1,3,9,10,11)], timevar = "query", idvar = c("term_id","source","term_name"), direction = "wide")
colnames(pwsPD1.r) <- gsub("p_value.","",colnames(pwsPD1.r))
write.csv(pwsPD1.r, "./Manuscript_Figures/figure5/DOWN_GO_ORA_Mfuzz_PD1.csv")
saveRDS(CLUSTERlistPD1, "./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_PD1.rds")

pwsPD1_plot <- pwsPD1
pwsPD1_plot$query <- paste0("cluster",pwsPD1_plot$query)

#selection
to_select <- pwsPD1_plot %>% 
  arrange(desc(p_value)) %>% 
  group_by(query) %>% dplyr::slice(1:15) %>%
  pull(term_name) %>%
  unique(.)
pwsPD1_plot <- pwsPD1_plot %>% dplyr::filter(term_name %in% to_select)
pwsPD1_plot <- pwsPD1_plot %>% distinct(term_name, query, .keep_all = TRUE)
midpoint <- (max(pwsPD1_plot$p_value) + min(pwsPD1_plot$p_value))/ 2

# # Ordering plot
pwsPD1_plot$query <- as.factor(pwsPD1_plot$query)
pwsPD1_plot <- pwsPD1_plot[order(pwsPD1_plot$query,
                                 pwsPD1_plot$p_value),]
term_order <- pwsPD1_plot$term_name %>% unique(.)
pwsPD1_plot$term_name <- factor(pwsPD1_plot$term_name,
                                levels = term_order)
#plot
svg(paste("./Manuscript_Figures/figure5/PathwayDownregulated_PD1_MFuzz.svg", sep=""), 
    width = 14, height = 14 )
ggplot(pwsPD1_plot, aes(x = query, y = term_name)) + 
  geom_point(aes(fill = p_value), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "p_value")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "Gene enrichment for downregulated stimulated\n PD1 genes accross MFuzz clusters")
dev.off()

#################
#### Heatmap ####
#################
CLUSTERlistPD1 <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistPD1 <- split(CLUSTERlistPD1, CLUSTERlistPD1$cluster)

for(i in 1:length(CLUSTERlistPD1)){CLUSTERlistPD1[[i]] <- CLUSTERlistPD1[[i]] %>%
  dplyr::select(-cluster) %>%
  dplyr::pull(gene)
} 


## 4 hours
library(ComplexHeatmap)
pat <- "(_4h)"
stim_design_4h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("TCR 4h",3), rep("PD1 4h",3), 
                             rep("BTLA 4h",3) ))
type$group <- factor(type$group, levels = c("TCR 4h", "PD1 4h", "BTLA 4h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("TCR 4h" = "#198482", "PD1 4h" = "#F49C40", 
                                            "BTLA 4h" = "red")))


library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)

for (i in 1:length(CLUSTERlistPD1)){
  temp <- normalized_counts[CLUSTERlistPD1[[i]], stim_design_4h]
  
  temp <- temp[apply(temp[,-1], 1, function(x) !all(x==0)),]
  
  svg(paste0("./Manuscript_Figures/others/Heatmap_downregulated_PD1__MFuzz_Cluster_4h_",i,".svg"), 
      width = 7, height = 4.5 )
  print(ComplexHeatmap::Heatmap(matrix = (temp %>% pheatmap:::scale_rows()), 
                                col = col,top_annotation = ha1, 
                                show_heatmap_legend = T, show_row_names = F, 
                                column_title = paste0("Cluster ",i," genes 4 hours"), name = " "))
  dev.off()
}


## 20 hours
library(ComplexHeatmap)
pat <- "(_20h)"
stim_design_20h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("TCR 20h",3), rep("PD1 20h",3), 
                             rep("BTLA 20h",3) ))
type$group <- factor(type$group, levels = c("TCR 20h", "PD1 20h", "BTLA 20h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("TCR 20h" = "#198482", "PD1 20h" = "#F49C40", 
                                            "BTLA 20h" = "red")))


library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)

for (i in 1:length(CLUSTERlistPD1)){
  temp <- normalized_counts[CLUSTERlistPD1[[i]], stim_design_20h]
  
  temp <- temp[apply(temp[,-1], 1, function(x) !all(x==0)),]
  
  svg(paste0("./Manuscript_Figures/others/Heatmap_downregulated_PD1__MFuzz_Cluster_20h_",i,".svg"), 
      width = 7, height = 4.5 )
  print(ComplexHeatmap::Heatmap(matrix = (temp %>% pheatmap:::scale_rows()), 
                                col = col,top_annotation = ha1, 
                                show_heatmap_legend = T, show_row_names = F, 
                                column_title = paste0("Cluster ",i," genes 20 hours"), name = " "))
  dev.off()
}

## 48 hours
library(ComplexHeatmap)
pat <- "(_48h)"
stim_design_48h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("TCR 48h",3), rep("PD1 48h",3), 
                             rep("BTLA 48h",3) ))
type$group <- factor(type$group, levels = c("TCR 48h", "PD1 48h", "BTLA 48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("TCR 48h" = "#198482", "PD1 48h" = "#F49C40", 
                                            "BTLA 48h" = "red")))


library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)

for (i in 1:length(CLUSTERlistPD1)){
  temp <- normalized_counts[CLUSTERlistPD1[[i]], stim_design_48h]
  
  temp <- temp[apply(temp[,-1], 1, function(x) !all(x==0)),]
  
  svg(paste0("./Manuscript_Figures/others/Heatmap_downregulated_PD1__MFuzz_Cluster_48h_",i,".svg"), 
      width = 7, height = 4.5 )
  print(ComplexHeatmap::Heatmap(matrix = (temp %>% pheatmap:::scale_rows()), 
                                col = col,top_annotation = ha1, 
                                show_heatmap_legend = T, show_row_names = F, 
                                column_title = paste0("Cluster ",i," genes 48 hours"), name = " "))
  dev.off()
}




#### Motif analysis ####
CLUSTERlistPD1 <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistPD1 <- split(CLUSTERlistPD1, CLUSTERlistPD1$cluster)

for(i in 1:length(CLUSTERlistPD1)){
  CLUSTERlistPD1[[i]] <- CLUSTERlistPD1[[i]] %>% 
    dplyr::select(-cluster) %>% pull(.)
} 
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

motif_PD1_MFuzz_up <- as.list(rep(NA, length(CLUSTERlistPD1)))
for(i in 1:length(motif_PD1_MFuzz_up)){
  tryCatch(
    expr = {
      temp <- intersect(CLUSTERlistPD1[[i]], colnames(motifRankings))
      motif_PD1_MFuzz_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                           motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_PD1_MFuzz_up[[i]] <- NULL
    })
}
names(motif_PD1_MFuzz_up) <- paste0("cluster",1:length(motif_PD1_MFuzz_up))
for (i in 1:length(motif_PD1_MFuzz_up)){
  if(length(motif_PD1_MFuzz_up[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_PD1_MFuzz_up[[i]] <- motif_PD1_MFuzz_up[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_PD1_MFuzz_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_PD1_MFuzz_up[[i]]$TF_highConf)
    motif_PD1_MFuzz_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_PD1_MFuzz_up[[i]]$TF_highConf)
    motif_PD1_MFuzz_up[[i]]$TF_highConf <- sub(" .*", "", motif_PD1_MFuzz_up[[i]]$TF_highConf)
    motif_PD1_MFuzz_up[[i]]$TF_highConf <- gsub(";", "", motif_PD1_MFuzz_up[[i]]$TF_highConf)
    motif_PD1_MFuzz_up[[i]]$group <- names(motif_PD1_MFuzz_up)[i]
  }
}

motif_PD1_MFuzz_up <- motif_PD1_MFuzz_up[lapply(motif_PD1_MFuzz_up,length)>1] ## you can use sapply,rapply
motif_PD1_MFuzz_up_combined <- do.call(rbind, motif_PD1_MFuzz_up)
motif_PD1_MFuzz_up_combined <- motif_PD1_MFuzz_up_combined %>% unique(.)

#selection
to_select_PD1_MFuzz_up <- motif_PD1_MFuzz_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:15) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_PD1_MFuzz_up_combined <- motif_PD1_MFuzz_up_combined %>% dplyr::filter(TF_highConf %in% to_select_PD1_MFuzz_up)
motif_PD1_MFuzz_up_combined <- motif_PD1_MFuzz_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_PD1_up <- (max(motif_PD1_MFuzz_up_combined$NES) + min(motif_PD1_MFuzz_up_combined$NES))/ 2

# # Ordering plot
motif_PD1_MFuzz_up_combined$group <- as.factor(motif_PD1_MFuzz_up_combined$group)
motif_PD1_MFuzz_up_combined <- motif_PD1_MFuzz_up_combined[order(motif_PD1_MFuzz_up_combined$group,
                                                                 motif_PD1_MFuzz_up_combined$NES),]
TF_order <- motif_PD1_MFuzz_up_combined$TF_highConf %>% unique(.)
motif_PD1_MFuzz_up_combined$TF_highConf <- factor(motif_PD1_MFuzz_up_combined$TF_highConf,
                                                  levels = TF_order)
#plot
motif_PD1_up_plot <- ggplot(motif_PD1_MFuzz_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_PD1_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for downregulated stimulated PD1 genes accross MFuzz clusters")

svg(paste("./Manuscript_Figures/figure5/MotifDownregulated_PD1_MFuzz.svg", sep=""), 
    width = 28, height = 7 )
motif_PD1_up_plot
dev.off()


################################
## Kinase Enrichment Analysis ##
################################
#TCR
DEGs.selection.up <- CLUSTERlistPD1

KEA_results <- as.list(rep(NA, length(DEGs.selection.up)))
names(KEA_results) <- names(DEGs.selection.up)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_PD1_MFuzz.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_PD1_MFuzz.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- 1:4

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = 1:4 )
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = 1:4)
type$group <- factor(type$group, levels = 1:4)


to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- 1:4
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA
colnames(new_df) <- paste("Cluster", colnames(new_df), sep = "_")

svg(paste0("./Manuscript_Figures/figure5//KEA_PD1_Heatmap_Mfuzz.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun,
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("PD1: Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()





#### Stimulated T-cells sc scoring ####

## UCell ## 
library(UCell)
CLUSTERlistPD1 <- readRDS("./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_PD1.rds")

for (i in 1:length(CLUSTERlistPD1)){
  cluster_MFuzz_PD1 <-  rbind(CLUSTERlistPD1[[i]] )%>% 
    pull(.)
  
  activated_seurat_Blood <- AddModuleScore_UCell(
    obj  = activated_seurat_Blood,
    features = list(cluster_MFuzz_PD1), name = paste0("cluster",i)
  )
}
plot_PD1_downregulated <- as.list(rep(NA, length(CLUSTERlistPD1)))
for (i in 1:length(CLUSTERlistPD1)){
  plot_PD1_downregulated[[i]] <- FeaturePlot(activated_seurat_Blood, 
                                             features = paste0("signature_1cluster",i) )+
    scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))+
    labs(title = paste("PD1 Signature MFuzz Cluster",i))
  
}
library(cowplot)
svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_UMAP_PD1.svg", sep=""), 
    width = 14, height = 14 )
plot_grid(plot_PD1_downregulated[[1]],plot_PD1_downregulated[[2]],
          plot_PD1_downregulated[[3]],plot_PD1_downregulated[[4]],
          labels = NULL)
dev.off()

## Seurat ##
CLUSTERlistPD1 <- readRDS("./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_PD1.rds")

for (i in 1:length(CLUSTERlistPD1)){
  cluster_MFuzz_PD1 <-  rbind(CLUSTERlistPD1[[i]] )%>% 
    pull(.)
  
  activated_seurat_Blood <- AddModuleScore(
    object  = activated_seurat_Blood,
    features = list(cluster_MFuzz_PD1), name= paste0("cluster",i)
  )
}
plot_PD1_downregulated <- as.list(rep(NA, length(CLUSTERlistPD1)))
for (i in 1:length(CLUSTERlistPD1)){
  plot_PD1_downregulated[[i]] <- FeaturePlot(activated_seurat_Blood, 
                                             features = paste0("cluster",i,"1") )+
    scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))+
    labs(title = paste("PD1 Signature MFuzz Cluster",i))
  
}
library(cowplot)
svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_UMAP_PD1_Seurat.svg", sep=""), 
    width = 14, height = 14 )
plot_grid(plot_PD1_downregulated[[1]],plot_PD1_downregulated[[2]],
          plot_PD1_downregulated[[3]],plot_PD1_downregulated[[4]],
          labels = NULL)
dev.off()




##########
## BTLA ##
##########

#####################
### Downregulated ###
#####################

#### MFuzz #### 
exp_design_BTLA <- exp_design %>% filter(str_detect(Remarks, c("BTLA_"))) %>% rownames(.)

DEGs_BTLA <- list()

DEGs_BTLA[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                            contrast = c("Remarks", "BTLA_4h", "ctrl_4h"), 
                                                            format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                             contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                             format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                             contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                             format = "DataFrame"))


for(i in 1:length(DEGs_BTLA)){
  DEGs_BTLA[[i]] <- DEGs_BTLA[[i]] %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -log2(1.5)) %>% 
    rownames()
}
DEGs_BTLA <- Reduce(c,DEGs_BTLA) %>% unique(.)
normalized_counts_BTLA_up <- normalized_counts[DEGs_BTLA, exp_design_BTLA]

# Aggregation and MFuzz clustering
normalized_counts_BTLA_up <- as.data.frame(t(normalized_counts_BTLA_up))
normalized_counts_BTLA_up$stim <- c(rep("BTLA 4h",3), 
                                    rep("BTLA 20h",3), rep("BTLA 48h", 3) )
normalized_counts_BTLA_up <- normalized_counts_BTLA_up%>%
  group_by(stim)%>% 
  summarise(across(everything(), list(mean)))
normalized_counts_BTLA_up <- normalized_counts_BTLA_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="stim")
normalized_counts_BTLA_up <- as.data.frame(t(normalized_counts_BTLA_up))
normalized_counts_BTLA_up$gene <- rownames(normalized_counts_BTLA_up)
normalized_counts_BTLA_up$gene <- gsub('.{2}$', '', normalized_counts_BTLA_up$gene)
normalized_counts_BTLA_up <- normalized_counts_BTLA_up %>% 
  remove_rownames %>% 
  column_to_rownames(var="gene")
col_order <- c( "BTLA 4h","BTLA 20h", "BTLA 48h")
normalized_counts_BTLA_up <- normalized_counts_BTLA_up[, col_order]

# TCR
exp_design_TCR_BTLA <- exp_design %>% filter(str_detect(Remarks, c("ctrl_"))) %>% rownames(.)
normalized_counts_TCR_BTLA <- normalized_counts[rownames(normalized_counts_BTLA_up), exp_design_TCR_BTLA]

# Aggregation and MFuzz clustering
normalized_counts_TCR_BTLA <- as.data.frame(t(normalized_counts_TCR_BTLA))
normalized_counts_TCR_BTLA$stim <- c(rep("stimulated 4h",3), 
                                     rep("stimulated 20h",3), rep("stimulated 48h", 3) )
normalized_counts_TCR_BTLA <- normalized_counts_TCR_BTLA%>%
  group_by(stim)%>% 
  summarise(across(everything(), list(mean)))
normalized_counts_TCR_BTLA <- normalized_counts_TCR_BTLA %>% 
  remove_rownames %>% 
  column_to_rownames(var="stim")
normalized_counts_TCR_BTLA <- as.data.frame(t(normalized_counts_TCR_BTLA))
normalized_counts_TCR_BTLA$gene <- rownames(normalized_counts_TCR_BTLA)
normalized_counts_TCR_BTLA$gene <- gsub('.{2}$', '', normalized_counts_TCR_BTLA$gene)
normalized_counts_TCR_BTLA <- normalized_counts_TCR_BTLA %>% 
  remove_rownames %>% 
  column_to_rownames(var="gene")
col_order <- c("stimulated 4h", "stimulated 20h", "stimulated 48h")
normalized_counts_TCR_BTLA <- normalized_counts_TCR_BTLA[, col_order]

# Substraction
normalized_counts_BTLA_up <- normalized_counts_BTLA_up - normalized_counts_TCR_BTLA



eset_BTLA <- new("ExpressionSet",exprs=as.matrix(normalized_counts_BTLA_up))

eset_BTLA <- Mfuzz::filter.NA(eset_BTLA, thres=0.25)
eset_BTLA <- Mfuzz::fill.NA(eset_BTLA,mode="mean")
eset_BTLA <- Mfuzz::standardise(eset_BTLA)
m1 <- mestimate(eset_BTLA)
n_clusters <- 4
col_order <- c( "BTLA 4h","BTLA 20h", "BTLA 48h")

set.seed(1)
cl_4 <- mfuzz(eset_BTLA,c=n_clusters,m=m1)
cl_4_BTLA <- mfuzz(eset_BTLA,c=n_clusters,m=m1)

svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_BTLA.svg", sep=""), 
    width = 14, height = 14 )
mfuzz.plot2(eset_BTLA,cl=cl_4,mfrow=c(2,2), x11 = FALSE , time.labels = col_order )
dev.off()
CLUSTERlistBTLA <- data.frame(cluster = cl_4_BTLA$cluster, gene = names(cl_4_BTLA$cluster))
CLUSTERlistBTLA <- split(CLUSTERlistBTLA, CLUSTERlistBTLA$cluster)


#KEA3 BTLA
for(i in 1:length(CLUSTERlistBTLA)){
  CLUSTERlistBTLA[[i]] <- CLUSTERlistBTLA[[i]]$gene
}

KEA_results <- as.list(rep(NA, length(CLUSTERlistBTLA)))
names(KEA_results) <- names(CLUSTERlistBTLA)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = CLUSTERlistBTLA[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_BTLA_MFuzz.xlsx")


# KEA visualization
top_10 <- lapply(KEA_results, function(i) head(i, 10)) 
for (i in 1:length(top_10)){top_10[[i]] <- top_10[[i]] %>% pull(TF)}
top_10 <- Reduce(c,top_10) %>% unique(.)
KEA_results_subset <- KEA_results


for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% dplyr::filter(TF %in% top_10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- paste0("cluster",i)
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$Score <- as.numeric(to_plot$Score)

pathway_order <- to_plot %>% dplyr::filter(group %in% "cluster1") %>%
  arrange(desc(logScore)) %>%
  pull(TF)
to_plot$TF <- factor(to_plot$TF, levels = pathway_order )
to_plot$logScore <- ifelse(to_plot$Score < 100, to_plot$logScore, NA)


KEA_bubble_plot <- ggplot(to_plot, aes(x = group, y = TF)) + 
  geom_point(aes(color = logScore ), size = 5) + 
  labs( x= "", y = "",size = "Log-score")  + 
  scale_color_gradient2(low="yellow", mid = "orange",high="red", midpoint = mean(to_plot$logScore, na.rm = T),na.value = "#FFFFFF" ) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", 
        axis.text = element_text(face = "bold", size = 16), 
        legend.text = element_text(face = "bold", size = 16), 
        legend.title = element_text(face = "bold", size = 16),
        title = element_text(face = "bold", size = 22) 
  )+
  labs(title = "Kinase Enrichment Analysis (BTLA)")

svg(paste("./Manuscript_Figures/figure5/KEA_BTLA_MFuzz_Plot.svg", sep=""), 
    width = 10, height = 10 )
KEA_bubble_plot
dev.off()



#### KEGG and REACTOME ORA enrichment analysis ####
CLUSTERlistBTLA <- data.frame(cluster = cl_4_BTLA$cluster, gene = names(cl_4_BTLA$cluster))
CLUSTERlistBTLA <- split(CLUSTERlistBTLA, CLUSTERlistBTLA$cluster)

for(i in 1:length(CLUSTERlistBTLA)){CLUSTERlistBTLA[[i]] <- CLUSTERlistBTLA[[i]] %>% dplyr::select(-cluster)}

pwsBTLA <- gost(CLUSTERlistBTLA, organism = "hsapiens", correction_method = "gSCS",
                user_threshold = 0.05, evcodes = TRUE, ordered_query = FALSE,
                sources = c("KEGG","REAC"))$result

pwsBTLA <- pwsBTLA[order(as.numeric(pwsBTLA$query)),]
pwsBTLA$p_value <- -log10(pwsBTLA$p_value) #transform to -log10(pval)
pwsBTLA.r <- reshape(pwsBTLA[,c(1,3,9,10,11)], timevar = "query", idvar = c("term_id","source","term_name"), direction = "wide")
colnames(pwsBTLA.r) <- gsub("p_value.","",colnames(pwsBTLA.r))
write.csv(pwsBTLA.r, "./Manuscript_Figures/figure5/DOWN_GO_ORA_Mfuzz_BTLA.csv")
saveRDS(CLUSTERlistBTLA, "./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_BTLA.rds")

pwsBTLA_plot <- pwsBTLA
pwsBTLA_plot$query <- paste0("cluster",pwsBTLA_plot$query)

#selection
to_select <- pwsBTLA_plot %>% 
  arrange(desc(p_value)) %>% 
  group_by(query) %>% dplyr::slice(1:15) %>%
  pull(term_name) %>%
  unique(.)
pwsBTLA_plot <- pwsBTLA_plot %>% dplyr::filter(term_name %in% to_select)
pwsBTLA_plot <- pwsBTLA_plot %>% distinct(term_name, query, .keep_all = TRUE)
midpoint <- (max(pwsBTLA_plot$p_value) + min(pwsBTLA_plot$p_value))/ 2

# # Ordering plot
pwsBTLA_plot$query <- as.factor(pwsBTLA_plot$query)
pwsBTLA_plot <- pwsBTLA_plot[order(pwsBTLA_plot$query,
                                   pwsBTLA_plot$p_value),]
term_order <- pwsBTLA_plot$term_name %>% unique(.)
pwsBTLA_plot$term_name <- factor(pwsBTLA_plot$term_name,
                                 levels = term_order)
#plot
svg(paste("./Manuscript_Figures/figure5/PathwayDownregulated_BTLA_MFuzz.svg", sep=""), 
    width = 14, height = 14 )
ggplot(pwsBTLA_plot, aes(x = query, y = term_name)) + 
  geom_point(aes(fill = p_value), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "p_value")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "Gene enrichment for downregulated stimulated\n BTLA genes accross MFuzz clusters")
dev.off()

#### Motif analysis ####
CLUSTERlistBTLA <- data.frame(cluster = cl_4$cluster, gene = names(cl_4$cluster))
CLUSTERlistBTLA <- split(CLUSTERlistBTLA, CLUSTERlistBTLA$cluster)

for(i in 1:length(CLUSTERlistBTLA)){
  CLUSTERlistBTLA[[i]] <- CLUSTERlistBTLA[[i]] %>% 
    dplyr::select(-cluster) %>% pull(.)
} 
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

motif_BTLA_MFuzz_up <- as.list(rep(NA, length(CLUSTERlistBTLA)))
for(i in 1:length(motif_BTLA_MFuzz_up)){
  tryCatch(
    expr = {
      temp <- intersect(CLUSTERlistBTLA[[i]], colnames(motifRankings))
      motif_BTLA_MFuzz_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                            motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_BTLA_MFuzz_up[[i]] <- NULL
    })
}
names(motif_BTLA_MFuzz_up) <- paste0("cluster",1:length(motif_BTLA_MFuzz_up))
for (i in 1:length(motif_BTLA_MFuzz_up)){
  if(length(motif_BTLA_MFuzz_up[[i]]) == 1 ){ # if empty
    print("no TF found")
  } else{
    motif_BTLA_MFuzz_up[[i]] <- motif_BTLA_MFuzz_up[[i]] %>%
      filter(str_detect(TF_highConf, "directAnnotation")) %>%
      dplyr::select(TF_highConf, NES)
    motif_BTLA_MFuzz_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_BTLA_MFuzz_up[[i]]$TF_highConf)
    motif_BTLA_MFuzz_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_BTLA_MFuzz_up[[i]]$TF_highConf)
    motif_BTLA_MFuzz_up[[i]]$TF_highConf <- sub(" .*", "", motif_BTLA_MFuzz_up[[i]]$TF_highConf)
    motif_BTLA_MFuzz_up[[i]]$TF_highConf <- gsub(";", "", motif_BTLA_MFuzz_up[[i]]$TF_highConf)
    motif_BTLA_MFuzz_up[[i]]$group <- names(motif_BTLA_MFuzz_up)[i]
  }
}

motif_BTLA_MFuzz_up <- motif_BTLA_MFuzz_up[lapply(motif_BTLA_MFuzz_up,length)>1] ## you can use sapply,rapply
motif_BTLA_MFuzz_up_combined <- do.call(rbind, motif_BTLA_MFuzz_up)
motif_BTLA_MFuzz_up_combined <- motif_BTLA_MFuzz_up_combined %>% unique(.)

#selection
to_select_BTLA_MFuzz_up <- motif_BTLA_MFuzz_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:15) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_BTLA_MFuzz_up_combined <- motif_BTLA_MFuzz_up_combined %>% dplyr::filter(TF_highConf %in% to_select_BTLA_MFuzz_up)
motif_BTLA_MFuzz_up_combined <- motif_BTLA_MFuzz_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_BTLA_up <- (max(motif_BTLA_MFuzz_up_combined$NES) + min(motif_BTLA_MFuzz_up_combined$NES))/ 2

# # Ordering plot
motif_BTLA_MFuzz_up_combined$group <- as.factor(motif_BTLA_MFuzz_up_combined$group)
motif_BTLA_MFuzz_up_combined <- motif_BTLA_MFuzz_up_combined[order(motif_BTLA_MFuzz_up_combined$group,
                                                                   motif_BTLA_MFuzz_up_combined$NES),]
TF_order <- motif_BTLA_MFuzz_up_combined$TF_highConf %>% unique(.)
motif_BTLA_MFuzz_up_combined$TF_highConf <- factor(motif_BTLA_MFuzz_up_combined$TF_highConf,
                                                   levels = TF_order)
#plot
motif_BTLA_up_plot <- ggplot(motif_BTLA_MFuzz_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 9) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_BTLA_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold")) +
  labs(title = "High confidence transcription factor for downregulated stimulated BTLA genes accross MFuzz clusters")

svg(paste("./Manuscript_Figures/figure5/MotifDownregulated_BTLA_MFuzz.svg", sep=""), 
    width = 28, height = 7 )
motif_BTLA_up_plot
dev.off()


################################
## Kinase Enrichment Analysis ##
################################
#TCR
DEGs.selection.up <- CLUSTERlistBTLA

KEA_results <- as.list(rep(NA, length(DEGs.selection.up)))
names(KEA_results) <- names(DEGs.selection.up)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 1:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_BTLA_MFuzz.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_BTLA_MFuzz.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- 1:4

for (i in 1:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = 1:4 )
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = 1:4)
type$group <- factor(type$group, levels = 1:4)


to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- 1:4
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA
colnames(new_df) <- paste("Cluster", colnames(new_df), sep = "_")

svg(paste0("./Manuscript_Figures/figure5//KEA_BTLA_Heatmap_Mfuzz.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun,
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("BTLA: Kinase Enrichment Score (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()






#### Stimulated T-cells sc scoring ####
library(UCell)
CLUSTERlistBTLA <- readRDS("./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_BTLA.rds")

for (i in 1:length(CLUSTERlistBTLA)){
  cluster_MFuzz_BTLA <-  rbind(CLUSTERlistBTLA[[i]] )%>% 
    pull(.)
  
  activated_seurat_Blood <- AddModuleScore_UCell(
    obj  = activated_seurat_Blood,
    features = list(cluster_MFuzz_BTLA), name = paste0("cluster",i)
  )
}
plot_BTLA_downregulated <- as.list(rep(NA, length(CLUSTERlistBTLA)))
for (i in 1:length(CLUSTERlistBTLA)){
  plot_BTLA_downregulated[[i]] <- FeaturePlot(activated_seurat_Blood, 
                                              features = paste0("signature_1cluster",i) )+
    scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))+
    labs(title = paste("BTLA Signature MFuzz Cluster",i))
  
}
library(cowplot)
svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_UMAP_BTLA.svg", sep=""), 
    width = 14, height = 14 )
plot_grid(plot_BTLA_downregulated[[1]],plot_BTLA_downregulated[[2]],
          plot_BTLA_downregulated[[3]],plot_BTLA_downregulated[[4]],
          labels = NULL)
dev.off()

## Seurat ##
CLUSTERlistBTLA <- readRDS("./Manuscript_Figures/figure5/DOWN_Mfuzz_genes_BTLA.rds")

for (i in 1:length(CLUSTERlistBTLA)){
  cluster_MFuzz_BTLA <-  rbind(CLUSTERlistBTLA[[i]] )%>% 
    pull(.)
  
  activated_seurat_Blood <- AddModuleScore(
    object  = activated_seurat_Blood,
    features = list(cluster_MFuzz_BTLA), name= paste0("cluster",i)
  )
}
plot_BTLA_downregulated <- as.list(rep(NA, length(CLUSTERlistBTLA)))
for (i in 1:length(CLUSTERlistBTLA)){
  plot_BTLA_downregulated[[i]] <- FeaturePlot(activated_seurat_Blood, 
                                              features = paste0("cluster",i,"1") )+
    scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))+
    labs(title = paste("BTLA Signature MFuzz Cluster",i))
  
}
library(cowplot)
svg(paste("./Manuscript_Figures/figure5/DOWN_Mfuzz_UMAP_BTLA_Seurat.svg", sep=""), 
    width = 14, height = 14 )
plot_grid(plot_BTLA_downregulated[[1]],plot_BTLA_downregulated[[2]],
          plot_BTLA_downregulated[[3]],plot_BTLA_downregulated[[4]],
          labels = NULL)
dev.off()

```


## TCR DEGs relation to PD1 and BTLA

```{r}
#Upregulated
DEGs_TCR_Signif_up <- DEGs
for (i in 1:length(DEGs_TCR_Signif_up)){
  DEGs_TCR_Signif_up[[i]] <- DEGs_TCR_Signif_up[[i]] %>%
    dplyr::filter(padj < 0.05 & log2FoldChange > log2(1.5))
  colnames(DEGs_TCR_Signif_up[[i]]) <- paste("TCR", colnames(DEGs_TCR_Signif_up[[i]]), sep = "_")
  DEGs_TCR_Signif_up[[i]]$gene <- rownames(DEGs_TCR_Signif_up[[i]])
}

DEGs_PD1_up <- DEGs_PD1
for(i in 1:length(DEGs_PD1)){
  DEGs_PD1_up[[i]] <- DEGs_PD1_up[[i]][rownames(DEGs_TCR_Signif_up[[i]]),]
  DEGs_PD1_up[[i]]$direction <- ifelse(DEGs_PD1_up[[i]]$log2FoldChange > 0 , "upregulated", "downregulated")
  colnames(DEGs_PD1_up[[i]]) <- paste("PD1", colnames(DEGs_PD1_up[[i]]), sep = "_")
  DEGs_PD1_up[[i]]$gene <- rownames(DEGs_PD1_up[[i]])
}
DEGs_BTLA_up <- DEGs_BTLA
for(i in 1:length(DEGs_BTLA_up)){
  DEGs_BTLA_up[[i]] <- DEGs_BTLA_up[[i]][rownames(DEGs_TCR_Signif_up[[i]]),]
  DEGs_BTLA_up[[i]]$direction <- ifelse(DEGs_BTLA_up[[i]]$log2FoldChange > 0 , "upregulated", "downregulated")
  colnames(DEGs_BTLA_up[[i]]) <- paste("BTLA", colnames(DEGs_BTLA_up[[i]]), sep = "_")
  DEGs_BTLA_up[[i]]$gene <- rownames(DEGs_BTLA_up[[i]])
}
normalized_counts_up <- as.list(rep(NA, length(DEGs_TCR_Signif_up)))
exp_design3 <- exp_design
exp_design3$Remarks <- paste(rownames(exp_design3), exp_design3$Remarks, sep = "_")
for (i in 1:length(normalized_counts_up)){
  normalized_counts_up[[i]] <- normalized_counts
  colnames(normalized_counts_up[[i]]) <- exp_design3$Remarks
  normalized_counts_up[[i]] <- normalized_counts_up[[i]][rownames(DEGs_TCR_Signif_up[[i]]),]
  normalized_counts_up[[i]] <- as.data.frame(normalized_counts_up[[i]])
  normalized_counts_up[[i]]$gene <- rownames(normalized_counts_up[[i]])
}

# left join
combined_up <- as.list(rep(NA, length(DEGs_TCR_Signif_up)))
for (i in 1:length(combined_up)){
  combined_up[[i]] <- purrr::reduce(list(normalized_counts_up[[i]],DEGs_TCR_Signif_up[[i]],
                                         DEGs_PD1_up[[i]], DEGs_BTLA_up[[i]]), 
                                    dplyr::left_join, by = 'gene')
}
names(combined_up) <- c("4h TCR Upregulated", "20h TCR Upregulated", "48h TCR Upregulated")
write.xlsx(combined_up, "./Manuscript_Figures/others/TCR_DEGS_upregulated.xlsx")

motif_TCR_up_plot
```

# scRNA-seq DEG

```{r}
for (i in 1:length(activated_seurat_list)){
  print(
    FeaturePlot(object = activated_seurat_list[[i]], features = "NFKB1")
  )
  print(
    DimPlot(object = activated_seurat_list[[i]], group.by = "State")
  )
}


Idents(activated_seurat_Blood) <- 'celltype'

de_markers_GammDelta <- FindMarkers(activated_seurat_Blood, 
                                    ident.1 = "-T cells", ident.2 = NULL, only.pos = TRUE)
de_markers_GammDelta <- arrange(de_markers_GammDelta, desc(avg_log2FC))
de_markers_GammDelta$gene <- rownames(de_markers_GammDelta)

for (i in 1:10){
  print(
    FeaturePlot(activated_seurat_Blood, features = rownames(de_markers_GammDelta)[i])
  )
}
```

# supplementary

```{r}
library(ggpubr)

exp_design_df <- exp_design
exp_design_df$sample <- rownames(exp_design_df)

### PDCD1 and BTLA expression 
temp <- normalized_counts[c("PDCD1","BTLA"),]
temp_exp_design <- exp_design_df
temp_exp_design$PDCD1 <- temp[1,]
temp_exp_design$BTLA <- temp[2,]

# PD1 and BTLA
pat <- "(ctrl|PD1|BTLA)"
temp_exp_design_PD1 <- temp_exp_design %>% filter(str_detect(Remarks, pat))
temp_exp_design_PD1 <- temp_exp_design_PD1 %>% separate(Remarks, c("stim", "time"), sep = "_", remove = F)
temp_exp_design_PD1$Remarks <- factor(temp_exp_design_PD1$Remarks, levels = c("ctrl_4h","PD1_4h","BTLA_4h",
                                                                              "ctrl_20h", "PD1_20h", "BTLA_20h",
                                                                              "ctrl_48h", "PD1_48h", "BTLA_48h"))
temp_exp_design_PD1$time <- factor(temp_exp_design_PD1$time, levels = c("4h","20h", "48h"))
temp_exp_design_PD1$stim <- factor(temp_exp_design_PD1$stim, levels = c("ctrl","PD1", "BTLA"))

temp_exp_design_PD1 <- temp_exp_design_PD1 %>% 
  pivot_longer(
    cols = `PDCD1`:`BTLA`, 
    names_to = "gene",
    values_to = "expression"
)
bp <- ggbarplot(
  temp_exp_design_PD1, x = "Remarks", y = "expression", fill = "gene",
  palette = "npg", add = "mean_sd",
  position = position_dodge(0.8)
  )+
  scale_fill_manual(values = c("ctrl" = "#cccccc",
                                "PDCD1" = "#00aa00",
                                "BTLA" = "#c837ab")) +
  labs( x = "") +
  theme(text = element_text(size = 30))


svg(paste("./Manuscript_Figures/supp/PDCD1_BTLA_expression.svg", sep=""), 
    width = 20, height = 6 )
bp 
dev.off()


### Not TCR DEG, but in PD1 and BTLA
TCR_Unregulated <- as.list(rep(NA, length(DEGs)))
DEGs_PD1_UnstimTCR <- as.list(rep(NA, length(DEGs)))
DEGs_BTLA_UnstimTCR <- as.list(rep(NA, length(DEGs)))

for(i in 1:length(TCR_Unregulated)){
  TCR_Unregulated[[i]] <- DEGs[[i]] %>% 
    dplyr::filter(abs(log2FoldChange) < log2(1.5) | padj > 0.05)
}

TCR_Unregulated <- Reduce(intersect, list(rownames(TCR_Unregulated[[1]]),
                                          rownames(TCR_Unregulated[[2]]),
                                          rownames(TCR_Unregulated[[3]])))

for(i in 1:length(DEGs_PD1)){
  DEGs_PD1_UnstimTCR[[i]] <- DEGs_PD1[[i]] %>% 
    dplyr::filter(row.names(DEGs_PD1[[i]]) %in% TCR_Unregulated )%>%
    dplyr::filter(abs(log2FoldChange) > log2(1.5) & padj < 0.05)
  DEGs_BTLA_UnstimTCR[[i]] <- DEGs_BTLA[[i]] %>% 
    dplyr::filter(row.names(DEGs_BTLA[[i]]) %in% TCR_Unregulated ) %>%
    dplyr::filter(abs(log2FoldChange) > log2(1.5) & padj < 0.05)
}
names(DEGs_PD1_UnstimTCR) <- c("4h","20h","48h")
names(DEGs_BTLA_UnstimTCR) <- c("4h", "20h", "48h")
write.xlsx(DEGs_PD1_UnstimTCR, "Manuscript_Figures/supp/PD1_UnstimTCR.xlsx", rowNames = T)
write.xlsx(DEGs_BTLA_UnstimTCR, "Manuscript_Figures/supp/BTLA_UnstimTCR.xlsx", rowNames = T)

### All genes ranked by stat heatmap
library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)

DEGs_ranked <- DEGs
for(i in 1:length(DEGs)){
  DEGs_ranked[[i]] <- setorder(DEGs[[i]],-stat)
}
#4h
pat <- "(unstim|_4h)"
stim_design_4h <- exp_design %>% filter(str_detect(Remarks, pat))
normalized_counts_4h <- normalized_counts[rownames(DEGs_ranked[[2]]),rownames(stim_design_4h)]

type <- data.frame(group = c(rep("Unstimulated",3), rep("TCR 4h",3), rep("PD1 4h",3), 
                             rep("BTLA 4h",3) ))
type$group <- factor(type$group, levels = c("Unstimulated", "TCR 4h", "PD1 4h", "BTLA 4h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("Unstimulated" = "black","TCR 4h" = "#cccccc", "PD1 4h" = "#00aa00", "BTLA 4h" = "#c838ab")))

normalized_counts_4h[normalized_counts_4h == 0] <- NA
svg(paste("./Manuscript_Figures/supp/Heatmap_all_4h.svg", sep=""), 
    width = 15, height = 35 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_4h %>% pheatmap:::scale_rows() ), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        cluster_columns = F, cluster_rows = F,na_col = "black",
                        column_title = "4 hours", name = " ", column_title_gp = gpar(fontsize = 25, fontface = "bold"))
dev.off()

#20h
pat <- "(unstim|_20h)"
stim_design_20h <- exp_design %>% filter(str_detect(Remarks, pat))
normalized_counts_20h <- normalized_counts[rownames(DEGs_ranked[[2]]),rownames(stim_design_20h)]

type <- data.frame(group = c(rep("Unstimulated",3), rep("TCR 20h",3), rep("PD1 20h",3), 
                             rep("BTLA 20h",3) ))
type$group <- factor(type$group, levels = c("Unstimulated", "TCR 20h", "PD1 20h", "BTLA 20h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("Unstimulated" = "black","TCR 20h" = "#cccccc", "PD1 20h" = "#00aa00", "BTLA 20h" = "#c838ab")))

normalized_counts_20h[normalized_counts_20h == 0] <- NA
svg(paste("./Manuscript_Figures/supp/Heatmap_all_20h.svg", sep=""), 
    width = 15, height = 35 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_20h %>% pheatmap:::scale_rows() ), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        cluster_columns = F, cluster_rows = F,na_col = "black",
                        column_title = "20 hours", name = " ", column_title_gp = gpar(fontsize = 25, fontface = "bold"))
dev.off()

#48h
pat <- "(unstim|_48h)"
stim_design_48h <- exp_design %>% filter(str_detect(Remarks, pat))
normalized_counts_48h <- normalized_counts[rownames(DEGs_ranked[[2]]),rownames(stim_design_48h)]

type <- data.frame(group = c(rep("Unstimulated",3), rep("TCR 48h",3), rep("PD1 48h",3), 
                             rep("BTLA 48h",3) ))
type$group <- factor(type$group, levels = c("Unstimulated", "TCR 48h", "PD1 48h", "BTLA 48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("Unstimulated" = "black","TCR 48h" = "#cccccc", "PD1 48h" = "#00aa00", "BTLA 48h" = "#c838ab")))

normalized_counts_48h[normalized_counts_48h == 0] <- NA
svg(paste("./Manuscript_Figures/supp/Heatmap_all_48h.svg", sep=""), 
    width = 15, height = 35 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_48h %>% pheatmap:::scale_rows() ), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        cluster_columns = F, cluster_rows = F,na_col = "black",
                        column_title = "48 hours", name = " ", column_title_gp = gpar(fontsize = 25, fontface = "bold"))
dev.off()


to_write <- list("4h" = normalized_counts_4h, "20h" = normalized_counts_20h, 
                 "48h" = normalized_counts_48h)
write.xlsx(x = to_write, 
           file = "Manuscript_Figures/supp/all_genes_ranked.xlsx",rowNames = TRUE)
# Investigate E2F targets and G2M checkpoint 
# PD1
DEGs_stat_PD1 <- NULL
for(i in 1:length(DEGs_PD1)){
  DEGs_stat_PD1[[i]] <- DEGs_PD1[[i]]$stat
  names(DEGs_stat_PD1[[i]]) <- rownames(DEGs_PD1[[i]])
  DEGs_stat_PD1[[i]] <- sort(DEGs_stat_PD1[[i]], decreasing = T)
}
#Hallmark
fgseaRes_Hallmark_PD1 <- NULL
for(i in 1:length(DEGs_stat_PD1)){
  fgseaRes_Hallmark_PD1[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_PD1[[i]])
  fgseaRes_Hallmark_PD1[[i]] <- fgseaRes_Hallmark_PD1[[i]] %>% dplyr::filter(padj < 0.05)
  row.names(fgseaRes_Hallmark_PD1[[i]]) <- fgseaRes_Hallmark_PD1[[i]]$pathway
}


# BTLA
DEGs_stat_BTLA <- NULL
for(i in 1:length(DEGs_BTLA)){
  DEGs_stat_BTLA[[i]] <- DEGs_BTLA[[i]]$stat
  names(DEGs_stat_BTLA[[i]]) <- rownames(DEGs_BTLA[[i]])
  DEGs_stat_BTLA[[i]] <- sort(DEGs_stat_BTLA[[i]], decreasing = T)
}
#Hallmark
fgseaRes_Hallmark_BTLA <- NULL
for(i in 1:length(DEGs_stat_BTLA)){
  fgseaRes_Hallmark_BTLA[[i]] <- fgsea(pathways=pathwaysH, DEGs_stat_BTLA[[i]])
  fgseaRes_Hallmark_BTLA[[i]] <- fgseaRes_Hallmark_BTLA[[i]] %>% dplyr::filter(padj < 0.05)
  row.names(fgseaRes_Hallmark_BTLA[[i]]) <- fgseaRes_Hallmark_BTLA[[i]]$pathway
}



### 20 hours
test_BTLA <- fgseaRes_Hallmark_BTLA[[2]] %>% 
  dplyr::filter(pathway %in% "HALLMARK_G2M_CHECKPOINT") %>%
  dplyr::pull(leadingEdge)
test_PD1 <- fgseaRes_Hallmark_PD1[[2]] %>% 
  dplyr::filter(pathway %in% "HALLMARK_G2M_CHECKPOINT") %>%
  dplyr::pull(leadingEdge)

intersect_G2M <- setdiff(test_BTLA[[1]], test_PD1[[1]])
library(VennDiagram)

venn.diagram(
  x = list(test_BTLA[[1]], test_PD1[[1]]),
  category.names = c("BTLA" , "PD1"),
  filename = "Manuscript_Figures/supp/G2M_Genes_intersect_20h.png",
  output=T,
    # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 400,
  disable.logging = T
)

G2M_PD1 <- DEGs_PD1[[2]][test_PD1[[1]],]
G2M_BTLA <- DEGs_BTLA[[2]][test_BTLA[[1]],]

G2M_PD1$signif <- ifelse(G2M_PD1$padj < 0.05 & G2M_PD1$log2FoldChange > log2(1.5), 
                         "DEG", "non-DEG")
write.csv(G2M_PD1, "Manuscript_Figures/supp/G2M_genes_PD1_20hours.csv")
G2M_BTLA$signif <- ifelse(G2M_BTLA$padj < 0.05 & G2M_BTLA$log2FoldChange > log2(1.5), 
                         "DEG", "non-DEG")
write.csv(G2M_BTLA, "Manuscript_Figures/supp/G2M_genes_BTLA_20hours.csv")


### 48 hours
test_BTLA <- fgseaRes_Hallmark_BTLA[[3]] %>% 
  dplyr::filter(pathway %in% "HALLMARK_G2M_CHECKPOINT") %>%
  dplyr::pull(leadingEdge)
test_PD1 <- fgseaRes_Hallmark_PD1[[3]] %>% 
  dplyr::filter(pathway %in% "HALLMARK_G2M_CHECKPOINT") %>%
  dplyr::pull(leadingEdge)

intersect_G2M <- setdiff(test_BTLA[[1]], test_PD1[[1]])
library(VennDiagram)

venn.diagram(
  x = list(test_BTLA[[1]], test_PD1[[1]]),
  category.names = c("BTLA" , "PD1"),
  filename = "Manuscript_Figures/supp/G2M_Genes_intersect_48h.png",
  output=T,
  # Output features
  imagetype="png" ,
  height = 2300 , 
  width = 2300 , 
  resolution = 400,
  disable.logging = T
)

G2M_PD1 <- DEGs_PD1[[3]][test_PD1[[1]],]
G2M_BTLA <- DEGs_BTLA[[3]][test_BTLA[[1]],]

G2M_PD1$signif <- ifelse(G2M_PD1$padj < 0.05 & G2M_PD1$log2FoldChange > log2(1.5), 
                         "DEG", "non-DEG")
write.csv(G2M_PD1, "Manuscript_Figures/supp/G2M_genes_PD1_48hours.csv")
G2M_BTLA$signif <- ifelse(G2M_BTLA$padj < 0.05 & G2M_BTLA$log2FoldChange > log2(1.5), 
                          "DEG", "non-DEG")
write.csv(G2M_BTLA, "Manuscript_Figures/supp/G2M_genes_BTLA_48hours.csv")




## Heatmap with unstim
## 20 hours
library(ComplexHeatmap)
pat <- "(unstim|_20h)"
stim_design_20h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("unstimulated", 3), rep("TCR 20h",3), rep("PD1 20h",3), 
                             rep("BTLA 20h",3) ))
type$group <- factor(type$group, levels = c("unstimulated", "TCR 20h", "PD1 20h", "BTLA 20h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("unstimulated" = "lavenderblush", "TCR 20h" = "#cccccc", 
                                            "PD1 20h" = "#00aa00", "BTLA 20h" = "#c838ab")))



library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)


# upregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.up.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.PD1 <- lapply(DEGs.selection.up.PD1, rownames)

DEGs.selection.up.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.BTLA <- lapply(DEGs.selection.up.BTLA, rownames)

DEGs.selection.up.20h <- unique(c(DEGs.selection.up.PD1$PD1StimvsTCRstim_20, 
                                  DEGs.selection.up.BTLA$BTLAStimvsTCRstim_20))
DEGs.selection.up.20h <- intersect(DEGs.selection.up.20h, DEGs.TCR.up$StimvsUnstim_20)

normalized_counts_20h_up <- normalized_counts[DEGs.selection.up.20h, stim_design_20h]
normalized_counts_20h_up <- normalized_counts_20h_up[rowSums(
  !as.matrix(normalized_counts_20h_up)) < ncol(normalized_counts_20h_up), ]

svg(paste("./Manuscript_Figures/figure3/Heatmap_upregulated_20h_PD1_BTLA_Unstim.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_20h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes upregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 20 hour", name = " ")
dev.off()



# downregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.down.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.PD1 <- lapply(DEGs.selection.down.PD1, rownames)

DEGs.selection.down.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.BTLA <- lapply(DEGs.selection.down.BTLA, rownames)

DEGs.selection.down.20h <- unique(c(DEGs.selection.down.PD1$PD1StimvsTCRstim_20, 
                                    DEGs.selection.down.BTLA$BTLAStimvsTCRstim_20))
DEGs.selection.down.20h <- intersect(DEGs.selection.down.20h, DEGs.TCR.up$StimvsUnstim_20)

normalized_counts_20h_up <- normalized_counts[DEGs.selection.down.20h, stim_design_20h]
normalized_counts_20h_up <- normalized_counts_20h_up[rowSums(
  !as.matrix(normalized_counts_20h_up)) < ncol(normalized_counts_20h_up), ]



svg(paste("./Manuscript_Figures/figure3/Heatmap_downregulated_20h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_20h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes downregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 20 hour", name = " ")
dev.off()


## 48 hours
library(ComplexHeatmap)
pat <- "(unstim|_48h)"
stim_design_48h <- exp_design %>% filter(str_detect(Remarks, pat)) %>% rownames(.)

type <- data.frame(group = c(rep("unstimulated", 3), rep("TCR 48h",3), rep("PD1 48h",3), 
                             rep("BTLA 48h",3) ))
type$group <- factor(type$group, levels = c("unstimulated", "TCR 48h", "PD1 48h", "BTLA 48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("unstimulated" = "lavenderblush", "TCR 48h" = "#cccccc", 
                                            "PD1 48h" = "#00aa00", "BTLA 48h" = "#c838ab")))



library("RColorBrewer")
col <- colorRampPalette(brewer.pal(10, "RdYlBu"))(256)
col <- rev(col)


# upregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange > log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.up.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.PD1 <- lapply(DEGs.selection.up.PD1, rownames)

DEGs.selection.up.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.up.BTLA <- lapply(DEGs.selection.up.BTLA, rownames)

DEGs.selection.up.48h <- unique(c(DEGs.selection.up.PD1$PD1StimvsTCRstim_48, 
                                  DEGs.selection.up.BTLA$BTLAStimvsTCRstim_48))
DEGs.selection.up.48h <- intersect(DEGs.selection.up.48h, DEGs.TCR.up$StimvsUnstim_48)

normalized_counts_48h_up <- normalized_counts[DEGs.selection.up.48h, stim_design_48h]
normalized_counts_48h_up <- normalized_counts_48h_up[rowSums(
  !as.matrix(normalized_counts_48h_up)) < ncol(normalized_counts_48h_up), ]

svg(paste("./Manuscript_Figures/figure3/Heatmap_upregulated_48h_PD1_BTLA_Unstim.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_48h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes upregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 48 hour", name = " ")
dev.off()



# downregulated
DEGs.TCR.up <- lapply(DEGs, subset, padj < 0.05 & log2FoldChange < -log2(1.5))
DEGs.TCR.up <- lapply(DEGs.TCR.up, rownames)

DEGs.selection.down.PD1 <- lapply(DEGs_PD1, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.PD1 <- lapply(DEGs.selection.down.PD1, rownames)

DEGs.selection.down.BTLA <- lapply(DEGs_BTLA, subset, padj < 0.05 & abs(log2FoldChange) > log2(1.5))
DEGs.selection.down.BTLA <- lapply(DEGs.selection.down.BTLA, rownames)

DEGs.selection.down.48h <- unique(c(DEGs.selection.down.PD1$PD1StimvsTCRstim_48, 
                                    DEGs.selection.down.BTLA$BTLAStimvsTCRstim_48))
DEGs.selection.down.48h <- intersect(DEGs.selection.down.48h, DEGs.TCR.up$StimvsUnstim_48)

normalized_counts_48h_up <- normalized_counts[DEGs.selection.down.48h, stim_design_48h]
normalized_counts_48h_up <- normalized_counts_48h_up[rowSums(
  !as.matrix(normalized_counts_48h_up)) < ncol(normalized_counts_48h_up), ]



svg(paste("./Manuscript_Figures/figure3/Heatmap_downregulated_48h_PD1_BTLA.svg", sep=""), 
    width = 7, height = 4.5 )
ComplexHeatmap::Heatmap(matrix = (normalized_counts_48h_up %>% pheatmap:::scale_rows()), 
                        col = col,top_annotation = ha1, 
                        show_heatmap_legend = T, show_row_names = F, 
                        column_title = "Genes downregulated upon TCR stimulation\ninteresected with PD1 or BTLA DEGs at 48 hour", name = " ")
dev.off()


## Venn diagram all common DEGs PD1 and BTLA
DEGs_PD1 <- list()

DEGs_PD1[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))

DEGs_BTLA <- list()

DEGs_BTLA[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                          contrast = c("Remarks", "BTLA_4h", "ctrl_4h"), 
                                                          format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                           format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                           format = "DataFrame"))


DEGS_PD1_BTLA <- list()
DEGS_PD1_BTLA[["BTLAStimvsPD1stim_4"]] <- as.data.frame(results(dds, 
                                                          contrast = c("Remarks", "PD1_4h", "BTLA_4h"), 
                                                          format = "DataFrame"))
DEGS_PD1_BTLA[["BTLAStimvsPD1stim_20"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "PD1_20h", "BTLA_20h"), 
                                                           format = "DataFrame"))
DEGS_PD1_BTLA[["BTLAStimvsPD1stim_48"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "PD1_48h", "BTLA_48h"), 
                                                           format = "DataFrame"))



venn_list_PD1_down <- list()
venn_list_BTLA_down <- list()

for(i in 1:length(DEGs_PD1)){
  venn_list_PD1_down[[i]] <- DEGs_PD1[[i]] %>% 
    filter(padj < 0.05 & log2FoldChange < -log2(1.5)) %>%
    rownames(.)
}
names(venn_list_PD1_down) <- names(DEGs_PD1)
venn_list_PD1_down <- venn_list_PD1_down[2:3] #only 20 and 48 hours

for(i in 1:length(DEGs_BTLA)){
  venn_list_BTLA_down[[i]] <- DEGs_BTLA[[i]] %>% 
    filter(padj < 0.05 & log2FoldChange < -log2(1.5)) %>%
    rownames(.)
}
names(venn_list_BTLA_down) <- names(DEGs_BTLA)
venn_list_BTLA_down <- venn_list_BTLA_down[2:3] #only 20 and 48 hours

venn_down_combined <- c(venn_list_PD1_down, venn_list_BTLA_down)
names(venn_down_combined) <- c("PD1 20h", "PD1 48h", "BTLA 20h", "BTLA 48h")

myCol <- brewer.pal(4, "Pastel2")
venn.diagram(
        x = venn_down_combined,
        filename = "./Manuscript_Figures/supp/venn_DEGS_PD1_BTLA.png",
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 3700 , 
        width = 3700 , 
        resolution = 300,
        disable.logging = T,
        main = "PD1 and BTLA Downregulated DEGs",
        main.cex = 4,
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = 2,
        fontface = "bold",main.fontfamily = "arial",
        
        # Set names
        cat.cex = 3,
        cat.fontface = "bold",
        cat.fontfamily = "arial"
)


#Expression of CD69, CD27, EGR1, EGR3, NR4A3

temp <- normalized_counts[c("BATF", 
                            "CD69", "TNF", "EGR1", "EGR3", "NR4A3","ZFP36L1"),]
temp <- as.data.frame(t(temp))
temp$sample <- rownames(temp)
temp_exp_design <- exp_design_df
temp_exp_design <- left_join(temp_exp_design, temp)

# PD1 and BTLA
pat <- "(ctrl|PD1|BTLA)"
temp_exp_design_PD1 <- temp_exp_design %>% filter(str_detect(Remarks, pat))
temp_exp_design_PD1 <- temp_exp_design_PD1 %>% separate(Remarks, c("stim", "time"), sep = "_", remove = F)
temp_exp_design_PD1$Remarks <- factor(temp_exp_design_PD1$Remarks, levels = c("ctrl_4h","PD1_4h","BTLA_4h",
                                                                              "ctrl_20h", "PD1_20h", "BTLA_20h",
                                                                              "ctrl_48h", "PD1_48h", "BTLA_48h"))
temp_exp_design_PD1$time <- factor(temp_exp_design_PD1$time, levels = c("4h","20h", "48h"))
temp_exp_design_PD1$stim <- factor(temp_exp_design_PD1$stim, levels = c("ctrl","PD1", "BTLA"))

result <- as.list(rep(NA, (ncol(temp)-1) ))
names(result) <- colnames(temp)[1:(ncol(temp)-1)]



for (i in names(result)){
  temp_exp_design_PD1$temp <- temp_exp_design_PD1[,i]
  stat.test <- temp_exp_design_PD1 %>%
    group_by(time) %>%
    t_test(temp ~ stim, p.adjust.method = "hochberg")
  
  stat.test <- stat.test %>% dplyr::select(-c(p, p.adj, p.adj.signif)) #remove stat test from t-test, replace with DESeq2 instead
  
  stat_table <- data.frame()
  
  #create significance table
  for (j in 1:length(DEGS_PD1_BTLA)){
    temp_PD1 <- DEGs_PD1[[j]][i,]
    temp_PD1 <- temp_PD1 %>% dplyr::mutate(group1 = "ctrl", group2 = "PD1")
    temp_BTLA <-DEGs_BTLA[[j]][i,]
    temp_BTLA <- temp_BTLA %>% dplyr::mutate(group1 = "ctrl", group2 = "BTLA")
    temp_PD1_BTLA <-DEGS_PD1_BTLA[[j]][i,]
    temp_PD1_BTLA <- temp_PD1_BTLA %>% dplyr::mutate(group1 = "PD1", group2 = "BTLA")
    
    temp_df <- rbind(temp_PD1, temp_BTLA, temp_PD1_BTLA)
    temp_df <- temp_df %>% dplyr::select(group1, group2, padj, log2FoldChange)
    temp_df$time <- ifelse(j == 1, "4h", ifelse(j == 2, "20h", "48h"))
    
    stat_table <- rbind(stat_table, temp_df)
  }
  
  stat.test <- left_join(stat.test, stat_table, by=c('group1', 'group2', "time"))
  stat.test$p.adj.signif <- ifelse(stat.test$padj < 0.05 & stat.test$padj > 0.005, "*", 
                                   ifelse(stat.test$padj < 0.005 & stat.test$padj > 0.0005, "**", 
                                          ifelse(stat.test$padj < 0.0005, "***", "ns")))

  bp <- ggbarplot(
    temp_exp_design_PD1, x = "time", y = "temp", fill = "stim",
    palette = "npg", add = "mean_sd",
    position = position_dodge(0.8)
    )+
    scale_fill_manual(values = c("ctrl" = "#cccccc",
                                  "PD1" = "#00aa00",
                                  "BTLA" = "#c837ab")) +
    labs(x = "", y = "", title = paste0(as.character(i)," Expression"))+
    theme(text = element_text(size = 25))
  
  stat.test <- stat.test %>%
    add_xy_position(x = "time", fun = "mean_sd", dodge = 1, step.increase = 0.5)
  
  temp2 <-   bp + 
    stat_pvalue_manual(
      stat.test, label = "p.adj.signif", tip.length = 0.05
      )
  result[[i]] <- temp2
  
  # svg(paste0("./Manuscript_Figures/supp/", i,"_expression.svg"), 
  #     width = 10, height = 10 )
  # bp + 
  #   stat_pvalue_manual(
  #     stat.test, label = "p.adj.signif", tip.length = 0.05
  #     )
  # dev.off()
}
library(cowplot)

svg("./Manuscript_Figures/supp/selected_genes_expression.svg", width = 20, height = 10 )
print(plot_grid(result[[1]], result[[2]], result[[3]], result[[4]], result[[5]], result[[6]], 
          labels = NULL, label_size = 12))
dev.off()



## Motif analysis upregulated PD1 and BTLA DEGs
library(RcisTarget)
# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)
motifRankings <- importRankings("/mnt/8TB/Projects/POIAZ/Donagh/SCENIC_10k_genes/data/hg19-tss-centered-10kb-7species.mc9nr.feather")

# Upregulated PD1
DEGs_signif <- DEGs_PD1
for (i in 1:length(DEGs_signif)){DEGs_signif[[i]] <- DEGs[[i]] %>% dplyr::filter(padj < 0.05)}
DEGs_signif_up <- DEGs_signif
for (i in 1:length(DEGs_signif_up)){
  DEGs_signif_up[[i]] <- DEGs_signif_up[[i]] %>% 
    dplyr::filter(log2FoldChange > log2(1.5)) %>% rownames(.)
}

motif_TCR_up <- as.list(rep(NA, length(DEGs_signif_up)))
for(i in 1:length(motif_TCR_up)){
    tryCatch(
      expr = {
        temp <- intersect(DEGs_signif_up[[i]], colnames(motifRankings))
        motif_TCR_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                         motifAnnot=motifAnnotations_hgnc)
      },
      error = function(e){
        message('insufficent genes')
        motif_TCR_up[[i]] <- NULL
      })
}
names(motif_TCR_up) <- c("PD1 4h", "PD1 20h", "BTLA 48h")
for (i in 1:length(motif_TCR_up)){
  motif_TCR_up[[i]] <- motif_TCR_up[[i]] %>%
    filter(str_detect(TF_highConf, "directAnnotation")) %>%
    dplyr::select(TF_highConf, NES)
  motif_TCR_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- sub(" .*", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub(";", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$group <- names(motif_TCR_up)[i]
}


motif_TCR_up_combined <- do.call(rbind, motif_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% unique(.)

#selection
to_select_TCR_up <- motif_TCR_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_TCR_up_combined <- motif_TCR_up_combined %>% dplyr::filter(TF_highConf %in% to_select_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_TCR_up <- (max(motif_TCR_up_combined$NES) + min(motif_TCR_up_combined$NES))/ 2

# Ordering plot
motif_TCR_up_combined$group <- factor(motif_TCR_up_combined$group, 
                                   levels = c("PD1 4h", "PD1 20h", "PD1 48h"))
motif_TCR_up_combined <- motif_TCR_up_combined[order(motif_TCR_up_combined$group,
                                                     motif_TCR_up_combined$NES),]
TF_order_TCR_up <- motif_TCR_up_combined$TF_highConf %>% unique(.)
motif_TCR_up_combined$TF_highConf <- factor(motif_TCR_up_combined$TF_highConf, 
                                   levels = TF_order_TCR_up)
#plot
motif_TCR_up_plot <- ggplot(motif_TCR_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_TCR_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold"),
        plot.title = element_text(size = 27, face = "bold")) +
  labs(title = "High confidence transcription factor for genes upregulated upon PD1 stimulation")
svg(paste("./Manuscript_Figures/supp/MotifUpregulated_PD1_NES2.svg", sep=""), 
    width = 24, height = 3 )
motif_TCR_up_plot
dev.off()

# Upregulated BTLA
DEGs_signif <- DEGs_BTLA
for (i in 1:length(DEGs_signif)){DEGs_signif[[i]] <- DEGs[[i]] %>% dplyr::filter(padj < 0.05)}
DEGs_signif_up <- DEGs_signif
for (i in 1:length(DEGs_signif_up)){
  DEGs_signif_up[[i]] <- DEGs_signif_up[[i]] %>% 
    dplyr::filter(log2FoldChange > log2(1.5)) %>% rownames(.)
}

motif_TCR_up <- as.list(rep(NA, length(DEGs_signif_up)))
for(i in 1:length(motif_TCR_up)){
  tryCatch(
    expr = {
      temp <- intersect(DEGs_signif_up[[i]], colnames(motifRankings))
      motif_TCR_up[[i]] <- cisTarget(temp, motifRankings,nesThreshold = 2,
                                     motifAnnot=motifAnnotations_hgnc)
    },
    error = function(e){
      message('insufficent genes')
      motif_TCR_up[[i]] <- NULL
    })
}
names(motif_TCR_up) <- c("BTLA 4h", "BTLA 20h", "BTLA 48h")
for (i in 1:length(motif_TCR_up)){
  motif_TCR_up[[i]] <- motif_TCR_up[[i]] %>%
    filter(str_detect(TF_highConf, "directAnnotation")) %>%
    dplyr::select(TF_highConf, NES)
  motif_TCR_up[[i]]$TF_highConf <-  gsub('\\.', "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub("\\ \\(directAnnotation\\)", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- sub(" .*", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$TF_highConf <- gsub(";", "", motif_TCR_up[[i]]$TF_highConf)
  motif_TCR_up[[i]]$group <- names(motif_TCR_up)[i]
}


motif_TCR_up_combined <- do.call(rbind, motif_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% unique(.)

#selection
to_select_TCR_up <- motif_TCR_up_combined %>% 
  arrange(NES) %>% 
  group_by(group) %>% dplyr::slice(1:30) %>%
  pull(TF_highConf) %>%
  unique(.)
motif_TCR_up_combined <- motif_TCR_up_combined %>% dplyr::filter(TF_highConf %in% to_select_TCR_up)
motif_TCR_up_combined <- motif_TCR_up_combined %>% distinct(TF_highConf, group, .keep_all = TRUE)
midpoint_TCR_up <- (max(motif_TCR_up_combined$NES) + min(motif_TCR_up_combined$NES))/ 2

# Ordering plot
motif_TCR_up_combined$group <- factor(motif_TCR_up_combined$group, 
                                      levels = c("BTLA 4h", "BTLA 20h", "BTLA 48h"))
motif_TCR_up_combined <- motif_TCR_up_combined[order(motif_TCR_up_combined$group,
                                                     motif_TCR_up_combined$NES),]
TF_order_TCR_up <- motif_TCR_up_combined$TF_highConf %>% unique(.)
motif_TCR_up_combined$TF_highConf <- factor(motif_TCR_up_combined$TF_highConf, 
                                            levels = TF_order_TCR_up)
#plot
motif_TCR_up_plot <- ggplot(motif_TCR_up_combined, aes(y = group, x = TF_highConf)) + 
  geom_point(aes(fill = NES), shape = 21, size = 5) + 
  labs( x= "", y = "", fill = "NES")  + 
  scale_fill_gradient2(low="yellow", mid = "orange",high="red", midpoint = midpoint_TCR_up) +
  theme_ipsum() +
  theme(legend.key=element_blank(), 
        # panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 19, face = "bold"),
        axis.text.y = element_text(size = 19, face = "bold"),
        legend.position = "right", legend.text = element_text(size = 15, face = "bold"),
        plot.title = element_text(size = 27, face = "bold")) +
  labs(title = "High confidence transcription factor for genes upregulated upon BTLA stimulation")
svg(paste("./Manuscript_Figures/supp/MotifUpregulated_BTLA_NES2.svg", sep=""), 
    width = 24, height = 3 )
motif_TCR_up_plot
dev.off()


#KEA for upregulated PD1 and BTLA genes

DEGs_PD1 <- list()

DEGs_PD1[["PD1StimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                  contrast = c("Remarks", "PD1_4h", "ctrl_4h"), 
                                                  format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_20h", "ctrl_20h"), 
                                                   format = "DataFrame"))
DEGs_PD1[["PD1StimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                   contrast = c("Remarks", "PD1_48h", "ctrl_48h"), 
                                                   format = "DataFrame"))

DEGs_BTLA <- list()

DEGs_BTLA[["BTLAStimvsTCRstim_4"]] <- as.data.frame(results(dds, 
                                                          contrast = c("Remarks", "BTLA_4h", "ctrl_4h"), 
                                                          format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_20"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_20h", "ctrl_20h"), 
                                                           format = "DataFrame"))
DEGs_BTLA[["BTLAStimvsTCRstim_48"]] <- as.data.frame(results(dds, 
                                                           contrast = c("Remarks", "BTLA_48h", "ctrl_48h"), 
                                                           format = "DataFrame"))



#PD1
DEGs.selection.up <- lapply(DEGs_PD1, subset, padj < 0.05 & log2FoldChange > log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 2:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_PD1.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_PD1.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("20h","48h")

for (i in 2:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

ha1 = HeatmapAnnotation(type = type$group,name = " ",
                        col = list(type = c("20h" = "#F49C40", 
                                            "48h" = "red")), which = "row", annotation_label = " ")

svg(paste0("./Manuscript_Figures/supp/KEA_PD1_Heatmap_Upregulated.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun, left_annotation =  ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score PD1 (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()



#BTLA
DEGs.selection.up <- lapply(DEGs_BTLA, subset, padj < 0.05 & log2FoldChange > log2(1.5))

for (i in 1:length(DEGs.selection.up)){
  DEGs.selection.up[[i]] <- DEGs.selection.up[[i]] %>% 
    arrange(padj) %>%
    head(., 300)
}
DEGs.selection.up <- lapply(DEGs.selection.up, rownames)

KEA_results <- as.list(rep(NA, length(DEGs)))
names(KEA_results) <- names(DEGs)

url = "https://maayanlab.cloud/kea3/api/enrich/"
encode = "json"
for (i in 2:length(KEA_results)){
  genes = DEGs.selection.up[[i]]
  
  payload = list(query_name = "myQuery", gene_set = genes)
  
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  results = fromJSON(json)
  
  KEA_results[[i]] <- results$`Integrated--meanRank`
}

write.xlsx(KEA_results, "Manuscript_Figures/others/KEA3_BTLA.xlsx")
KEA_results <- read_excel_allsheets("Manuscript_Figures/others/KEA3_BTLA.xlsx")

KEA_results_subset <- KEA_results
type_TCR <- c("20h","48h")

for (i in 2:length(KEA_results)){
  KEA_results_subset[[i]] <- KEA_results[[i]] %>% head(10)
  KEA_results_subset[[i]]$logScore <- log(as.numeric(KEA_results_subset[[i]]$Score))
  KEA_results_subset[[i]]$group <- type_TCR[i]
}

to_plot <- bind_rows(KEA_results_subset)
to_plot$group <- factor(to_plot$group, levels = c("20h", "48h"))
to_plot$Score <- as.numeric(to_plot$Score)
to_plot$Rank <- as.numeric(to_plot$Rank)


library(ComplexHeatmap)
type <- data.frame(group = c(rep("20h",1), 
                             rep("48h",1) ))
type$group <- factor(type$group, levels = c("20h", "48h"))
ha1 = HeatmapAnnotation(type = type$group,
                        col = list(type = c("20h" = "#F49C40", 
                                            "48h" = "red")))

to_plot_heatmap <- to_plot %>% dplyr::select(TF, Rank, group)
to_plot_heatmap <- reshape(to_plot_heatmap, idvar = "TF", timevar = "group", direction = "wide")
to_plot_heatmap <- to_plot_heatmap %>% remove_rownames %>% column_to_rownames(var="TF")
colnames(to_plot_heatmap) <- c("20h", "48h")
to_plot_heatmap <- as.matrix(to_plot_heatmap)
to_plot_heatmap[is.na(to_plot_heatmap)] <- 0

library("RColorBrewer")
col_fun<-circlize::colorRamp2(c(1:10), colorRampPalette(brewer.pal(10, "YlOrRd"))(10)[10:1] )

gene_order <- ComplexHeatmap::Heatmap(matrix = (to_plot_heatmap), 
                                      col = col_fun,top_annotation = ha1, 
                                      show_heatmap_legend = T, show_row_names = T,
                                      na_col = "black",
                                      column_title = paste0("KEA3 Rank Score"), name = " ")
gene_order <- row_order(gene_order) %>% as.data.frame()
colnames(gene_order) <- "index"

to_plot_heatmap_dict <- data.frame("index" = 1:nrow(to_plot_heatmap),
                                   "gene" = rownames(to_plot_heatmap))
to_plot_heatmap_dict <- left_join(gene_order, to_plot_heatmap_dict)


new_df <- to_plot_heatmap[to_plot_heatmap_dict$index, ]
new_df[new_df == 0] <- NA

ha1 = HeatmapAnnotation(type = type$group,name = " ",
                        col = list(type = c("20h" = "#F49C40", 
                                            "48h" = "red")), which = "row", annotation_label = " ")

svg(paste0("./Manuscript_Figures/supp/KEA_BTLA_Heatmap_Upregulated.svg"), 
    width = 16, height = 4 )
print(ComplexHeatmap::Heatmap(matrix = t(new_df), 
                              col = col_fun, left_annotation =  ha1, 
                              show_heatmap_legend = T, show_row_names = T,
                              na_col = "grey",cluster_rows = F,cluster_columns = F,
                              column_title = paste0("Kinase Enrichment Score BTLA (Rank)"),
                              heatmap_legend_param = list(title = "Rank", at = 10:1),
                              row_names_side = "left",column_names_rot = 45,
                              name = "rank"))
dev.off()


# PD1
DEGs_stat_PD1 <- NULL
for(i in 1:length(DEGs_PD1)){
  DEGs_stat_PD1[[i]] <- DEGs_PD1[[i]]$stat
  names(DEGs_stat_PD1[[i]]) <- rownames(DEGs_PD1[[i]])
  DEGs_stat_PD1[[i]] <- sort(DEGs_stat_PD1[[i]], decreasing = T)
}
#Reactome
fgseaRes_Reactome <- NULL
for(i in 1:length(DEGs_stat_PD1)){
  fgseaRes_Reactome[[i]] <- fgsea(pathways=pathwaysC2_REACTOME, DEGs_stat_PD1[[i]])
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::filter(padj < 0.05)
}
pathways_Reactome <- Reduce(c, list(fgseaRes_Reactome[[1]]$pathway, fgseaRes_Reactome[[2]]$pathway, fgseaRes_Reactome[[3]]$pathway)) %>% 
  unique(.)
fgseaRes_Reactome <- vector(mode='list', length=3)
names(fgseaRes_Reactome) <- c("PD1 4h", "PD1 20h", "PD1 48h")
for(i in 1:length(DEGs_stat_PD1)){
  fgseaRes_Reactome[[i]] <- fgsea(pathways=pathwaysC2_REACTOME, DEGs_stat_PD1[[i]])
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::filter(pathway %in% pathways_Reactome)
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::select(pathway, padj, NES)
  fgseaRes_Reactome[[i]]$group <- names(fgseaRes_Reactome)[i]
}
fgseaRes_Reactome_plot <- do.call(rbind, fgseaRes_Reactome)
fgseaRes_Reactome_plot$group <- factor(fgseaRes_Reactome_plot$group, 
                                       levels = c("PD1 4h", "PD1 20h", "PD1 48h"))
fgseaRes_Reactome_plot$pathway <- gsub("Reactome_", "", fgseaRes_Reactome_plot$pathway)
fgseaRes_Reactome_plot$pathway <- gsub("_", " ", fgseaRes_Reactome_plot$pathway)

# Filtering. Too many results for Reactome
fgseaRes_Reactome_to_keep <- fgseaRes_Reactome_plot %>% 
  group_by(group) %>% 
  slice_min(order_by = padj, n = 20)
fgseaRes_Reactome_plot <- fgseaRes_Reactome_plot %>%
  dplyr::filter(pathway %in% fgseaRes_Reactome_to_keep$pathway)
fgseaRes_Reactome_plot$pathway <- gsub("REACTOME ", "", fgseaRes_Reactome_plot$pathway)

pathway_order <- fgseaRes_Reactome_plot %>% dplyr::filter(group %in% "PD1 20h") %>%
  arrange(NES) %>%
  pull(pathway)
fgseaRes_Reactome_plot$pathway <- factor(fgseaRes_Reactome_plot$pathway, 
                                         levels = pathway_order )
fgseaRes_Reactome_plot$padj <- ifelse(fgseaRes_Reactome_plot$padj > 0.05 , NA, fgseaRes_Reactome_plot$padj)
fgseaRes_Reactome_plot <- fgseaRes_Reactome_plot %>%
  group_by(pathway) %>%
  filter(!all(is.na(padj))) %>%
  ungroup()

Reactome_bubble_plot <- ggplot(fgseaRes_Reactome_plot, aes(x = group, y = pathway)) + 
  geom_point(aes(fill = NES, size = -log10(padj)), shape = 21) + 
  labs( x= "", y = "", fill = "NES", size = "-log10(padj)")  + 
  scale_fill_gradient2(low="blue", mid="white", high="red", oob = scales::squish) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", axis.text = element_text(face = "bold", size = 12), 
        legend.text = element_text(face = "bold", size = 10), 
        legend.title = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14) 
  )+
  labs(title = "Gene Set Enrichment Analysis (PD1)")


svg(paste("./Manuscript_Figures/figure3/ReactomePD1_Stim.svg", sep=""), 
    width = 18, height = 18 )
Reactome_bubble_plot
dev.off()

# BTLA
DEGs_stat_BTLA <- NULL
for(i in 1:length(DEGs_BTLA)){
  DEGs_stat_BTLA[[i]] <- DEGs_BTLA[[i]]$stat
  names(DEGs_stat_BTLA[[i]]) <- rownames(DEGs_BTLA[[i]])
  DEGs_stat_BTLA[[i]] <- sort(DEGs_stat_BTLA[[i]], decreasing = T)
}
#Reactome
fgseaRes_Reactome <- NULL
for(i in 1:length(DEGs_stat_BTLA)){
  fgseaRes_Reactome[[i]] <- fgsea(pathways=pathwaysC2_REACTOME, DEGs_stat_BTLA[[i]])
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::filter(padj < 0.05)
}
pathways_Reactome <- Reduce(c, list(fgseaRes_Reactome[[1]]$pathway, fgseaRes_Reactome[[2]]$pathway, fgseaRes_Reactome[[3]]$pathway)) %>% 
  unique(.)
fgseaRes_Reactome <- vector(mode='list', length=3)
names(fgseaRes_Reactome) <- c("BTLA 4h", "BTLA 20h", "BTLA 48h")
for(i in 1:length(DEGs_stat_BTLA)){
  fgseaRes_Reactome[[i]] <- fgsea(pathways=pathwaysC2_REACTOME, DEGs_stat_BTLA[[i]])
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::filter(pathway %in% pathways_Reactome)
  fgseaRes_Reactome[[i]] <- fgseaRes_Reactome[[i]] %>% dplyr::select(pathway, padj, NES)
  fgseaRes_Reactome[[i]]$group <- names(fgseaRes_Reactome)[i]
}
fgseaRes_Reactome_plot <- do.call(rbind, fgseaRes_Reactome)
fgseaRes_Reactome_plot$group <- factor(fgseaRes_Reactome_plot$group, 
                                       levels = c("BTLA 4h", "BTLA 20h", "BTLA 48h"))
fgseaRes_Reactome_plot$pathway <- gsub("Reactome_", "", fgseaRes_Reactome_plot$pathway)
fgseaRes_Reactome_plot$pathway <- gsub("_", " ", fgseaRes_Reactome_plot$pathway)

# Filtering. Too many results for Reactome
fgseaRes_Reactome_to_keep <- fgseaRes_Reactome_plot %>% 
  group_by(group) %>% 
  slice_min(order_by = padj, n = 20)
fgseaRes_Reactome_plot <- fgseaRes_Reactome_plot %>%
  dplyr::filter(pathway %in% fgseaRes_Reactome_to_keep$pathway)
fgseaRes_Reactome_plot$pathway <- gsub("REACTOME ", "", fgseaRes_Reactome_plot$pathway)


pathway_order <- fgseaRes_Reactome_plot %>% dplyr::filter(group %in% "BTLA 20h") %>%
  arrange(NES) %>%
  pull(pathway)
fgseaRes_Reactome_plot$pathway <- factor(fgseaRes_Reactome_plot$pathway, 
                                         levels = pathway_order )
fgseaRes_Reactome_plot$padj <- ifelse(fgseaRes_Reactome_plot$padj > 0.05 , NA, fgseaRes_Reactome_plot$padj)
fgseaRes_Reactome_plot <- fgseaRes_Reactome_plot %>%
  group_by(pathway) %>%
  filter(!all(is.na(padj))) %>%
  ungroup()

Reactome_bubble_plot <- ggplot(fgseaRes_Reactome_plot, aes(x = group, y = pathway)) + 
  geom_point(aes(fill = NES, size = -log10(padj)), shape = 21) + 
  labs( x= "", y = "", fill = "NES", size = "-log10(padj)")  + 
  scale_fill_gradient2(low="blue", mid="white", high="red", oob = scales::squish) +
  theme(legend.key=element_blank(), 
        panel.background = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2), 
        legend.position = "right", axis.text = element_text(face = "bold", size = 12), 
        legend.text = element_text(face = "bold", size = 10), 
        legend.title = element_text(face = "bold", size = 10),
        title = element_text(face = "bold", size = 14) 
  )+
  labs(title = "Gene Set Enrichment Analysis (BTLA)")



svg(paste("./Manuscript_Figures/figure3/ReactomeBTLA_Stim.svg", sep=""), 
    width = 18, height = 18 )
Reactome_bubble_plot
dev.off()
```

